{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "aaa23ad0-6f98-48e1-adc5-8d676fbd6320",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/Documents/programming/mele_lab/projects/240706_pantranscriptome_cerberus_gtf_merge/scripts/\n"
     ]
    }
   ],
   "source": [
    "import pyranges as pr\n",
    "import seaborn as sns\n",
    "import yaml\n",
    "import os\n",
    "import sys\n",
    "from snakemake.io import expand\n",
    "\n",
    "p = os.path.dirname(os.path.dirname(os.path.dirname(os.getcwd())))+'/scripts/'\n",
    "print(p)\n",
    "sys.path.append(p)\n",
    "\n",
    "from sm_utils import *\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "13124ee4-30cf-4358-8d64-bd1cd3823202",
   "metadata": {},
   "outputs": [],
   "source": [
    "config = load_config()\n",
    "od = '../../../'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a1fb08b8-82c2-4498-93aa-92744870f5ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis = 'espresso_pseudomasked_genomic'\n",
    "tool = 'espresso'\n",
    "config_tsv = od+f'snakemake/config_{analysis}_expression.tsv'\n",
    "meta_df = parse_config(config_tsv)\n",
    "meta_df['analysis'] = analysis\n",
    "tech_rep = ['GM10493_1']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2a55708-8e9f-4705-83c6-296e75eda795",
   "metadata": {},
   "source": [
    "## Assign novel genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "31f01a61-f07d-4045-bd64-a73b4e6c5727",
   "metadata": {},
   "outputs": [],
   "source": [
    "ifile = od+f'data/{analysis}/raw/2_PY2_GM10493.gtf'\n",
    "bed = expand(od+config['fmt']['novel_gene_merge_bed'], \n",
    "       analysis=analysis)[0]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "8103ee4e-9bbe-4f72-ab76-6d3616dd3774",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_df = rename_novel_genes(ifile, bed, 'test', tool)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "f3f81ecd-6e78-44cd-9e87-0da0f91d1dbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rename_novel_genes(ifile, bed, ofile, tool):\n",
    "    \"\"\"\n",
    "    Rename novel genes based on the name of overlapping bed regions\n",
    "    in a GTF file\n",
    "    \"\"\"\n",
    "    \n",
    "    # get entries w/ novel GIDs\n",
    "    gtf_df = pr.read_gtf(ifile, as_df=True)\n",
    "    nov_gids = get_gtf_novel_genes(gtf_df, tool)\n",
    "    nov_df = gtf_df.loc[gtf_df.gene_id.isin(nov_gids)]\n",
    "    l1 = len(nov_df.index)\n",
    "\n",
    "    # use tid or gid based on tool\n",
    "    if tool == 'iq':\n",
    "        id_col = 'gene_id'\n",
    "    elif tool == 'espresso':\n",
    "        id_col = 'transcript_id'\n",
    "\n",
    "    # merge w/ bed\n",
    "    df = pr.read_bed(bed)\n",
    "    nov_df = pr.PyRanges(nov_df)\n",
    "    nov_df = nov_df.join(df,\n",
    "                          how='left',\n",
    "                          strandedness='same',\n",
    "                          preserve_order=True)\n",
    "\n",
    "    # sanity checks\n",
    "    nov_df = nov_df.df\n",
    "    l2 = len(nov_df.index)\n",
    "    assert l1 == l2\n",
    "    assert len(nov_df.loc[nov_df.Name.isnull()].index) == 0\n",
    "\n",
    "\n",
    "    # get a dict to map sample-level novel gene ids to cross-sample novel gene ids\n",
    "    id_dict = dict([(o,n) for o,n in zip(nov_df[id_col].tolist(),\n",
    "                                         nov_df['Name'].tolist())])\n",
    "    nov_df[[id_col, 'Name']].drop_duplicates().head()\n",
    "\n",
    "    # now update full gtf\n",
    "    nov_ids = nov_df[id_col].unique().tolist()\n",
    "    inds = gtf_df.loc[gtf_df[id_col].isin(nov_ids)].index\n",
    "    gtf_df.loc[inds, 'gene_id'] = gtf_df.loc[inds, id_col].map(id_dict)\n",
    "    \n",
    "    return gtf_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "fd1b0c36-57bf-4196-b30d-58b3719df7a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get entries w/ novel GIDs\n",
    "gtf_df = pr.read_gtf(ifile, as_df=True)\n",
    "nov_gids = get_gtf_novel_genes(gtf_df, tool)\n",
    "nov_df = gtf_df.loc[gtf_df.gene_id.isin(nov_gids)]\n",
    "l1 = len(nov_df.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "872cb761-9b55-4cfb-91aa-f1c98caec371",
   "metadata": {},
   "outputs": [],
   "source": [
    "# use tid or gid based on tool\n",
    "if tool == 'iq':\n",
    "    id_col = 'gene_id'\n",
    "elif tool == 'espresso':\n",
    "    id_col = 'transcript_id'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "33a0a9f2-5c12-47b6-9d6c-0cfb3c384666",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge w/ bed\n",
    "df = pr.read_bed(bed)\n",
    "nov_df = pr.PyRanges(nov_df)\n",
    "nov_df = nov_df.join(df,\n",
    "                      how='left',\n",
    "                      strandedness='same',\n",
    "                      preserve_order=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "c20fb6cf-6c4a-4a1f-8feb-0882fbb9bb4a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score_b</th>\n",
       "      <th>Strand_b</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Chromosome, Source, Feature, Start, End, Score, Strand, Frame, transcript_id, gene_id, exon_number, Start_b, End_b, Name, Score_b, Strand_b]\n",
       "Index: []"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nov_df = nov_df.df\n",
    "nov_df.loc[nov_df.Name.isnull()]\n",
    "nov_df.loc[nov_df.Name == 'NA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "80cb20ba-b302-48f2-9a37-6ae32bb23e00",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get a dict to map sample-level novel gene ids to cross-sample novel gene ids\n",
    "id_dict = dict([(o,n) for o,n in zip(nov_df[id_col].tolist(),\n",
    "                                     nov_df['Name'].tolist())])\n",
    "nov_df[[id_col, 'Name']].drop_duplicates().head()\n",
    "\n",
    "# now update full gtf\n",
    "nov_ids = nov_df[id_col].unique().tolist()\n",
    "inds = gtf_df.loc[gtf_df[id_col].isin(nov_ids)].index\n",
    "gtf_df.loc[inds, 'gene_id'] = gtf_df.loc[inds, id_col].map(id_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "c0b06832-7cd5-4b0b-a9d3-a341c75e5d9c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1360\n"
     ]
    }
   ],
   "source": [
    "l  = gtf_df.loc[(gtf_df.gene_id.str.contains('novel'))].gene_id.unique().tolist()\n",
    "print(len(l))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "45252e9a-c2c8-4f65-bcc7-e27ccbe3b64f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>exon_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>213606</th>\n",
       "      <td>GL000205.2</td>\n",
       "      <td>novel_isoform</td>\n",
       "      <td>transcript</td>\n",
       "      <td>137791</td>\n",
       "      <td>145075</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ESPRESSO:GL000205.2:9:0</td>\n",
       "      <td>novel_gene_4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213607</th>\n",
       "      <td>GL000205.2</td>\n",
       "      <td>novel_isoform</td>\n",
       "      <td>exon</td>\n",
       "      <td>137791</td>\n",
       "      <td>138168</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ESPRESSO:GL000205.2:9:0</td>\n",
       "      <td>novel_gene_4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213608</th>\n",
       "      <td>GL000205.2</td>\n",
       "      <td>novel_isoform</td>\n",
       "      <td>exon</td>\n",
       "      <td>144905</td>\n",
       "      <td>145075</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ESPRESSO:GL000205.2:9:0</td>\n",
       "      <td>novel_gene_4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome         Source     Feature   Start     End Score Strand  \\\n",
       "213606  GL000205.2  novel_isoform  transcript  137791  145075     .      -   \n",
       "213607  GL000205.2  novel_isoform        exon  137791  138168     .      -   \n",
       "213608  GL000205.2  novel_isoform        exon  144905  145075     .      -   \n",
       "\n",
       "       Frame            transcript_id       gene_id exon_number  \n",
       "213606     .  ESPRESSO:GL000205.2:9:0  novel_gene_4         NaN  \n",
       "213607     .  ESPRESSO:GL000205.2:9:0  novel_gene_4           1  \n",
       "213608     .  ESPRESSO:GL000205.2:9:0  novel_gene_4           2  "
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_df.loc[gtf_df[id_col] == 'ESPRESSO:GL000205.2:9:0']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45b405b5-3768-41b7-a463-e8f05f942c8f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2281b54c-7704-41d2-ad58-8b90cb6b7f12",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02588cd3-7742-40cd-b254-e4187daf6401",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a215dbd0-e689-4553-bc42-b73c3a9a99b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_hier_entry(df, how='t'):\n",
    "    \"\"\"\n",
    "    kind {'g','t'}\n",
    "    \"\"\"\n",
    "    agg_dict = {'min_coord': 'min', 'max_coord': 'max'}\n",
    "    t_df = df.copy(deep=True)\n",
    "    t_df['min_coord'] = t_df[['Start', 'End']].min(axis=1)\n",
    "    t_df['max_coord'] = t_df[['Start', 'End']].max(axis=1)\n",
    "    if how == 't':\n",
    "        gb_cols = ['Chromosome', 'Strand', 'gene_name',\n",
    "                   'gene_id', 'transcript_id', 'transcript_name',\n",
    "                   'tss_id', 'tes_id',\n",
    "                   'new_transcript_id', 'original_transcript_id',\n",
    "                   'original_transcript_name', 'ag1', 'ag2']\n",
    "    elif how == 'g':\n",
    "        gb_cols = ['Chromosome', 'Strand', 'gene_name',\n",
    "                   'gene_id']\n",
    "    gb_cols = list(set(gb_cols)&(set(t_df.columns)))\n",
    "\n",
    "    cols = gb_cols + ['min_coord', 'max_coord']\n",
    "    t_df = t_df[cols]\n",
    "    t_df = t_df.groupby(gb_cols, observed=True).agg(agg_dict).reset_index()\n",
    "    t_df.rename({'min_coord': 'Start', 'max_coord': 'End'}, axis=1, inplace=True)\n",
    "    if how == 't':\n",
    "        t_df['Feature'] = 'transcript'\n",
    "    elif how == 'g':\n",
    "        t_df['Feature'] = 'gene'\n",
    "\n",
    "    return t_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "038dbde6-114f-4d91-9027-20d3f7ea84a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pr.read_gtf('2_PY2_GM10493.gtf', as_df=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e5664bc9-1936-4603-a1b6-37f3a779137f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['NA']"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# noivel gene finder for esprsso\n",
    "novel_gids = df.loc[(df.Source=='novel_isoform')&\\\n",
    "                    (df.gene_id=='NA')].gene_id.unique().tolist()\n",
    "novel_gids = ['NA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "e81f40eb-723c-4700-9229-81976e2a9d08",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "986\n",
      "['ENSG00000273748.1', 'NA', 'ENSG00000115241.11', 'ENSG00000138002.16', 'ENSG00000171103.11']\n"
     ]
    }
   ],
   "source": [
    "l = df.loc[df.Source=='novel_isoform','gene_id'].unique().tolist()\n",
    "print(len(l))\n",
    "print(l[:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "4f5edc70-1412-493b-aa3f-4916784698ac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "215498\n",
      "12977\n",
      "13025\n"
     ]
    }
   ],
   "source": [
    "# TODO try on one gene\n",
    "# df = df.loc[df.gene_id=='ENSG00000296598.1']\n",
    "\n",
    "print(len(df.index))\n",
    "print(len(df.gene_id.unique()))\n",
    "\n",
    "df = make_hier_entry(df, how='g')\n",
    "print(len(df.loc[df.Feature=='gene'].index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "f1302320-bd63-4b59-a906-7df1b966c019",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Feature</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003.16</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX</td>\n",
       "      <td>100627107</td>\n",
       "      <td>100636806</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>-</td>\n",
       "      <td>chr20</td>\n",
       "      <td>50934869</td>\n",
       "      <td>50958532</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000457.14</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>169849630</td>\n",
       "      <td>169893959</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>169795042</td>\n",
       "      <td>169804347</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000938.13</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>27612063</td>\n",
       "      <td>27635185</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13020</th>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6</td>\n",
       "      <td>655943</td>\n",
       "      <td>167093593</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13021</th>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>chr7</td>\n",
       "      <td>716619</td>\n",
       "      <td>149226170</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13022</th>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>chr8</td>\n",
       "      <td>1482495</td>\n",
       "      <td>144012561</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13023</th>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>chr9</td>\n",
       "      <td>23296</td>\n",
       "      <td>137032119</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13024</th>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX</td>\n",
       "      <td>2888681</td>\n",
       "      <td>155264525</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13025 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  gene_id Strand Chromosome      Start        End Feature\n",
       "0      ENSG00000000003.16      -       chrX  100627107  100636806    gene\n",
       "1      ENSG00000000419.14      -      chr20   50934869   50958532    gene\n",
       "2      ENSG00000000457.14      -       chr1  169849630  169893959    gene\n",
       "3      ENSG00000000460.17      +       chr1  169795042  169804347    gene\n",
       "4      ENSG00000000938.13      -       chr1   27612063   27635185    gene\n",
       "...                   ...    ...        ...        ...        ...     ...\n",
       "13020                  NA      -       chr6     655943  167093593    gene\n",
       "13021                  NA      -       chr7     716619  149226170    gene\n",
       "13022                  NA      -       chr8    1482495  144012561    gene\n",
       "13023                  NA      -       chr9      23296  137032119    gene\n",
       "13024                  NA      -       chrX    2888681  155264525    gene\n",
       "\n",
       "[13025 rows x 6 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "42cccece-5e38-47b8-9b33-0d259e389461",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Feature</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12976</th>\n",
       "      <td>NA</td>\n",
       "      <td>+</td>\n",
       "      <td>GL000219.1</td>\n",
       "      <td>98559</td>\n",
       "      <td>99612</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13002</th>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>944205</td>\n",
       "      <td>244150826</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13003</th>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>chr10</td>\n",
       "      <td>511506</td>\n",
       "      <td>129864305</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13004</th>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>chr11</td>\n",
       "      <td>524290</td>\n",
       "      <td>125903200</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13005</th>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>chr12</td>\n",
       "      <td>567393</td>\n",
       "      <td>131074524</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13006</th>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>chr13</td>\n",
       "      <td>19674755</td>\n",
       "      <td>110728101</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13007</th>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>chr14</td>\n",
       "      <td>20659870</td>\n",
       "      <td>106844796</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13008</th>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>chr15</td>\n",
       "      <td>22166998</td>\n",
       "      <td>99896391</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13009</th>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>chr16</td>\n",
       "      <td>46989</td>\n",
       "      <td>88983277</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13010</th>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17</td>\n",
       "      <td>400747</td>\n",
       "      <td>82212841</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13011</th>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>chr18</td>\n",
       "      <td>3247880</td>\n",
       "      <td>80202195</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13012</th>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>chr19</td>\n",
       "      <td>1026607</td>\n",
       "      <td>58499293</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13013</th>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>chr2</td>\n",
       "      <td>10883878</td>\n",
       "      <td>241553603</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13014</th>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>chr20</td>\n",
       "      <td>320334</td>\n",
       "      <td>58458506</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13015</th>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>chr21</td>\n",
       "      <td>14516408</td>\n",
       "      <td>45615831</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13016</th>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>chr22</td>\n",
       "      <td>16667616</td>\n",
       "      <td>49657520</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13017</th>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>chr3</td>\n",
       "      <td>204459</td>\n",
       "      <td>193593109</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13018</th>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>chr4</td>\n",
       "      <td>485561</td>\n",
       "      <td>185209453</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13019</th>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>chr5</td>\n",
       "      <td>1488302</td>\n",
       "      <td>180808778</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13020</th>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6</td>\n",
       "      <td>655943</td>\n",
       "      <td>167093593</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13021</th>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>chr7</td>\n",
       "      <td>716619</td>\n",
       "      <td>149226170</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13022</th>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>chr8</td>\n",
       "      <td>1482495</td>\n",
       "      <td>144012561</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13001</th>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>GL000205.2</td>\n",
       "      <td>137791</td>\n",
       "      <td>145075</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13023</th>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>chr9</td>\n",
       "      <td>23296</td>\n",
       "      <td>137032119</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13000</th>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>GL000195.1</td>\n",
       "      <td>48901</td>\n",
       "      <td>86687</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12998</th>\n",
       "      <td>NA</td>\n",
       "      <td>+</td>\n",
       "      <td>chr9</td>\n",
       "      <td>1759160</td>\n",
       "      <td>136851017</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12977</th>\n",
       "      <td>NA</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1013497</td>\n",
       "      <td>248918646</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12978</th>\n",
       "      <td>NA</td>\n",
       "      <td>+</td>\n",
       "      <td>chr10</td>\n",
       "      <td>15097365</td>\n",
       "      <td>133391691</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12979</th>\n",
       "      <td>NA</td>\n",
       "      <td>+</td>\n",
       "      <td>chr11</td>\n",
       "      <td>9384614</td>\n",
       "      <td>124624361</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12980</th>\n",
       "      <td>NA</td>\n",
       "      <td>+</td>\n",
       "      <td>chr12</td>\n",
       "      <td>6960832</td>\n",
       "      <td>131739255</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12981</th>\n",
       "      <td>NA</td>\n",
       "      <td>+</td>\n",
       "      <td>chr13</td>\n",
       "      <td>23808310</td>\n",
       "      <td>109362230</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12982</th>\n",
       "      <td>NA</td>\n",
       "      <td>+</td>\n",
       "      <td>chr14</td>\n",
       "      <td>22533474</td>\n",
       "      <td>106864284</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12983</th>\n",
       "      <td>NA</td>\n",
       "      <td>+</td>\n",
       "      <td>chr15</td>\n",
       "      <td>31345429</td>\n",
       "      <td>101914500</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12984</th>\n",
       "      <td>NA</td>\n",
       "      <td>+</td>\n",
       "      <td>chr16</td>\n",
       "      <td>2009885</td>\n",
       "      <td>78757203</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12985</th>\n",
       "      <td>NA</td>\n",
       "      <td>+</td>\n",
       "      <td>chr17</td>\n",
       "      <td>16381343</td>\n",
       "      <td>82420559</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12986</th>\n",
       "      <td>NA</td>\n",
       "      <td>+</td>\n",
       "      <td>chr18</td>\n",
       "      <td>3247828</td>\n",
       "      <td>60884580</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12987</th>\n",
       "      <td>NA</td>\n",
       "      <td>+</td>\n",
       "      <td>chr19</td>\n",
       "      <td>1864296</td>\n",
       "      <td>52605468</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12988</th>\n",
       "      <td>NA</td>\n",
       "      <td>+</td>\n",
       "      <td>chr2</td>\n",
       "      <td>10897013</td>\n",
       "      <td>236414941</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12989</th>\n",
       "      <td>NA</td>\n",
       "      <td>+</td>\n",
       "      <td>chr20</td>\n",
       "      <td>7092535</td>\n",
       "      <td>57826018</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12990</th>\n",
       "      <td>NA</td>\n",
       "      <td>+</td>\n",
       "      <td>chr21</td>\n",
       "      <td>25467172</td>\n",
       "      <td>46313225</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12991</th>\n",
       "      <td>NA</td>\n",
       "      <td>+</td>\n",
       "      <td>chr22</td>\n",
       "      <td>17777328</td>\n",
       "      <td>42084285</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12992</th>\n",
       "      <td>NA</td>\n",
       "      <td>+</td>\n",
       "      <td>chr3</td>\n",
       "      <td>237272</td>\n",
       "      <td>186485925</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12993</th>\n",
       "      <td>NA</td>\n",
       "      <td>+</td>\n",
       "      <td>chr4</td>\n",
       "      <td>7754077</td>\n",
       "      <td>178658676</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12994</th>\n",
       "      <td>NA</td>\n",
       "      <td>+</td>\n",
       "      <td>chr5</td>\n",
       "      <td>4017393</td>\n",
       "      <td>180826497</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12995</th>\n",
       "      <td>NA</td>\n",
       "      <td>+</td>\n",
       "      <td>chr6</td>\n",
       "      <td>10412325</td>\n",
       "      <td>164525768</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12996</th>\n",
       "      <td>NA</td>\n",
       "      <td>+</td>\n",
       "      <td>chr7</td>\n",
       "      <td>602025</td>\n",
       "      <td>141288776</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12997</th>\n",
       "      <td>NA</td>\n",
       "      <td>+</td>\n",
       "      <td>chr8</td>\n",
       "      <td>435613</td>\n",
       "      <td>136731454</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12999</th>\n",
       "      <td>NA</td>\n",
       "      <td>+</td>\n",
       "      <td>chrX</td>\n",
       "      <td>6309241</td>\n",
       "      <td>153360101</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13024</th>\n",
       "      <td>NA</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX</td>\n",
       "      <td>2888681</td>\n",
       "      <td>155264525</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_id Strand  Chromosome     Start        End Feature\n",
       "12976      NA      +  GL000219.1     98559      99612    gene\n",
       "13002      NA      -        chr1    944205  244150826    gene\n",
       "13003      NA      -       chr10    511506  129864305    gene\n",
       "13004      NA      -       chr11    524290  125903200    gene\n",
       "13005      NA      -       chr12    567393  131074524    gene\n",
       "13006      NA      -       chr13  19674755  110728101    gene\n",
       "13007      NA      -       chr14  20659870  106844796    gene\n",
       "13008      NA      -       chr15  22166998   99896391    gene\n",
       "13009      NA      -       chr16     46989   88983277    gene\n",
       "13010      NA      -       chr17    400747   82212841    gene\n",
       "13011      NA      -       chr18   3247880   80202195    gene\n",
       "13012      NA      -       chr19   1026607   58499293    gene\n",
       "13013      NA      -        chr2  10883878  241553603    gene\n",
       "13014      NA      -       chr20    320334   58458506    gene\n",
       "13015      NA      -       chr21  14516408   45615831    gene\n",
       "13016      NA      -       chr22  16667616   49657520    gene\n",
       "13017      NA      -        chr3    204459  193593109    gene\n",
       "13018      NA      -        chr4    485561  185209453    gene\n",
       "13019      NA      -        chr5   1488302  180808778    gene\n",
       "13020      NA      -        chr6    655943  167093593    gene\n",
       "13021      NA      -        chr7    716619  149226170    gene\n",
       "13022      NA      -        chr8   1482495  144012561    gene\n",
       "13001      NA      -  GL000205.2    137791     145075    gene\n",
       "13023      NA      -        chr9     23296  137032119    gene\n",
       "13000      NA      -  GL000195.1     48901      86687    gene\n",
       "12998      NA      +        chr9   1759160  136851017    gene\n",
       "12977      NA      +        chr1   1013497  248918646    gene\n",
       "12978      NA      +       chr10  15097365  133391691    gene\n",
       "12979      NA      +       chr11   9384614  124624361    gene\n",
       "12980      NA      +       chr12   6960832  131739255    gene\n",
       "12981      NA      +       chr13  23808310  109362230    gene\n",
       "12982      NA      +       chr14  22533474  106864284    gene\n",
       "12983      NA      +       chr15  31345429  101914500    gene\n",
       "12984      NA      +       chr16   2009885   78757203    gene\n",
       "12985      NA      +       chr17  16381343   82420559    gene\n",
       "12986      NA      +       chr18   3247828   60884580    gene\n",
       "12987      NA      +       chr19   1864296   52605468    gene\n",
       "12988      NA      +        chr2  10897013  236414941    gene\n",
       "12989      NA      +       chr20   7092535   57826018    gene\n",
       "12990      NA      +       chr21  25467172   46313225    gene\n",
       "12991      NA      +       chr22  17777328   42084285    gene\n",
       "12992      NA      +        chr3    237272  186485925    gene\n",
       "12993      NA      +        chr4   7754077  178658676    gene\n",
       "12994      NA      +        chr5   4017393  180826497    gene\n",
       "12995      NA      +        chr6  10412325  164525768    gene\n",
       "12996      NA      +        chr7    602025  141288776    gene\n",
       "12997      NA      +        chr8    435613  136731454    gene\n",
       "12999      NA      +        chrX   6309241  153360101    gene\n",
       "13024      NA      -        chrX   2888681  155264525    gene"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.gene_id.duplicated(keep=False)].sort_values(by='gene_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc8ab2b9-c514-4119-8b94-791947ee5d3d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
