{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "24ffbdce-544d-45c7-9743-3d4bf5012634",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import cerberus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "111c71cd-281d-4b0b-9b1f-0658fff8cb1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_exon_df(n,c,e,s,g,t,nt=None,ag1=None,ag2=None):\n",
    "    df = pd.DataFrame()\n",
    "    df['Chromosome'] = c\n",
    "    try:\n",
    "        df['Start'] = [i[0] for i in e]\n",
    "    except:\n",
    "        import pdb; pdb.set_trace()\n",
    "    df['End'] = [i[1] for i in e]\n",
    "    df['Strand'] = s\n",
    "    df['Feature'] = 'exon'\n",
    "    cols = ['gene_name', 'gene_id']\n",
    "    for c in cols:\n",
    "        df[c] = g\n",
    "    cols = ['transcript_id', 'transcript_name']\n",
    "    for c in cols:\n",
    "        df[c] = t\n",
    "    if nt:\n",
    "        df['new_transcript_id'] = nt\n",
    "    if ag1:\n",
    "        df['ag1'] = ag1\n",
    "    if ag2:\n",
    "        df['ag2'] = ag2\n",
    "\n",
    "    # reorder exons and starts/ stops if needed\n",
    "    df['new_Start'] = df[['Start', 'End']].min(axis=1)\n",
    "    df['new_End'] = df[['Start', 'End']].max(axis=1)\n",
    "    df.drop(['Start', 'End'], axis=1, inplace=True)\n",
    "    df.rename({'new_Start':'Start',\n",
    "               'new_End':'End'}, axis=1, inplace=True)\n",
    "    s = s[0]\n",
    "    if s == '+':\n",
    "        ascending = True\n",
    "    elif s == '-':\n",
    "        ascending = False\n",
    "    df.sort_values(by='Start', ascending=ascending, inplace=True)\n",
    "    return df\n",
    "\n",
    "def make_hier_entry(df, how='t'):\n",
    "    \"\"\"\n",
    "    kind {'g','t'}\n",
    "    \"\"\"\n",
    "    agg_dict = {'min_coord': 'min', 'max_coord': 'max'}\n",
    "    t_df = df.copy(deep=True)\n",
    "    t_df['min_coord'] = t_df[['Start', 'End']].min(axis=1)\n",
    "    t_df['max_coord'] = t_df[['Start', 'End']].max(axis=1)\n",
    "    if how == 't':\n",
    "        gb_cols = ['Chromosome', 'Strand', 'gene_name',\n",
    "                   'gene_id', 'transcript_id', 'transcript_name',\n",
    "                   'tss_id', 'tes_id',\n",
    "                   'new_transcript_id', 'original_transcript_id',\n",
    "                   'original_transcript_name', 'ag1', 'ag2']\n",
    "        gb_cols = list(set(gb_cols)&(set(t_df.columns)))\n",
    "    elif how == 'g':\n",
    "        gb_cols = ['Chromosome', 'Strand', 'gene_name',\n",
    "                   'gene_id']\n",
    "\n",
    "    cols = gb_cols + ['min_coord', 'max_coord']\n",
    "    t_df = t_df[cols]\n",
    "    t_df = t_df.groupby(gb_cols).agg(agg_dict).reset_index()\n",
    "    t_df.rename({'min_coord': 'Start', 'max_coord': 'End'}, axis=1, inplace=True)\n",
    "    if how == 't':\n",
    "        t_df['Feature'] = 'transcript'\n",
    "    elif how == 'g':\n",
    "        t_df['Feature'] = 'gene'\n",
    "\n",
    "    return t_df\n",
    "\n",
    "def make_test_gtf(ts):\n",
    "    df = pd.concat(ts)\n",
    "        # make transcript entries\n",
    "    t_df = make_hier_entry(df, how='t')\n",
    "    # make gene entries\n",
    "    g_df = make_hier_entry(df, how='g')\n",
    "\n",
    "    # concat everything and sort by gene id, transcript id, feature rank (gene =0, t =1, exon=2), then start coords\n",
    "    df = pd.concat([df, t_df, g_df])\n",
    "    df = cerberus.sort_gtf(df)\n",
    "    return df\n",
    "\n",
    "\n",
    "# need:\n",
    "# transcripts that don't need to be aggregated\n",
    "# transcripts that do need to be aggregated\n",
    "\n",
    "ts = []\n",
    "\n",
    "# t1 - transcript that won't need to be aggregated\n",
    "e = [[1,10], [20,30]]\n",
    "n = len(e)\n",
    "c = ['chr1' for i in range(n)]\n",
    "s = ['+' for i in range(n)]\n",
    "g = 'g1'\n",
    "nt = 'g1[1,1,1]'\n",
    "t = 'g1_t1'\n",
    "ag1='known'\n",
    "ag2='p1'\n",
    "df = make_exon_df(n,c,e,s,g,t,nt,ag1,ag2)\n",
    "df['tss_id'] = 'g1_1'\n",
    "df['tes_id'] = 'g1_1'\n",
    "ts.append(df)\n",
    "\n",
    "\n",
    "# t3, g2\n",
    "e = [[1,55]]\n",
    "n = len(e)\n",
    "c = ['chr1' for i in range(n)]\n",
    "s = ['+' for i in range(n)]\n",
    "g = 'g2'\n",
    "e = [[x + 200 for x in e1] for e1 in e]\n",
    "nt = 'g2[1,1,3]'\n",
    "t = 'g2_t3'\n",
    "ag1 = 'novel'\n",
    "ag2 = 'p1'\n",
    "df = make_exon_df(n,c,e,s,g,t,nt,ag1,ag2)\n",
    "df['tss_id'] = 'g2_1'\n",
    "df['tes_id'] = 'g2_3'\n",
    "ts.append(df)\n",
    "\n",
    "# t5n-- 5/3' ends longer/ shorter, same IC\n",
    "e = [[20,30], [40,100]]\n",
    "n = len(e)\n",
    "c = ['chr1' for i in range(n)]\n",
    "s = ['+' for i in range(n)]\n",
    "g = 'g3'\n",
    "e = [[x + 300 for x in e1] for e1 in e]\n",
    "nt = 'g3[1,1,1]'\n",
    "t = 'g3_t5'\n",
    "ag1 = 'known'\n",
    "ag2 = 'p3'\n",
    "df = make_exon_df(n,c,e,s,g,t,nt,ag1,ag2)\n",
    "df['tss_id'] = 'g3_1'\n",
    "df['tes_id'] = 'g3_1'\n",
    "ts.append(df)\n",
    "\n",
    "# t7n-- 5/3' ends longer/ shorter, same IC\n",
    "e = [[20,30], [40,45]]\n",
    "n = len(e)\n",
    "c = ['chr1' for i in range(n)]\n",
    "s = ['+' for i in range(n)]\n",
    "g = 'g4'\n",
    "e = [[x + 400 for x in e1] for e1 in e]\n",
    "nt = 'g4[1,1,1]'\n",
    "t = 'g4_t7'\n",
    "ag1 = 'known'\n",
    "ag2 = 'p3'\n",
    "df = make_exon_df(n,c,e,s,g,t,nt,ag1,ag2)\n",
    "df['tss_id'] = 'g4_1'\n",
    "df['tes_id'] = 'g4_1'\n",
    "ts.append(df)\n",
    "\n",
    "# make test gtf\n",
    "test_df = make_test_gtf(ts)\n",
    "test_df.rename({'transcript_id': 'original_transcript_id',\n",
    "                'transcript_name': 'original_transcript_name'},\n",
    "               axis=1, inplace=True)\n",
    "test_df.rename({'new_transcript_id': 'transcript_id'},\n",
    "               axis=1, inplace=True)\n",
    "test_df['transcript_name'] = test_df['transcript_id']\n",
    "test_df['Source'] = 'test_1'\n",
    "test_df = pr.PyRanges(test_df)\n",
    "test_df.to_gtf('test_gtf_1.gtf')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "20251bc9-b03d-4f54-81e7-7846814ff36e",
   "metadata": {},
   "outputs": [],
   "source": [
    "ts = []\n",
    "# t2 - transcript with shorter 5'\n",
    "e = [[8,10], [20,30]]\n",
    "n = len(e)\n",
    "c = ['chr1' for i in range(n)]\n",
    "s = ['+' for i in range(n)]\n",
    "g = 'g1'\n",
    "nt = 'g1[2,1,1]'\n",
    "t = 'g1_t2'\n",
    "ag1='known'\n",
    "ag2='p1'\n",
    "df = make_exon_df(n,c,e,s,g,t,nt,ag1,ag2)\n",
    "df['tss_id'] = 'g1_1'\n",
    "df['tes_id'] = 'g1_1'\n",
    "ts.append(df)\n",
    "\n",
    "# t4, g2 monoexonic <80% overlap\n",
    "e = [[50,85]]\n",
    "n = len(e)\n",
    "c = ['chr1' for i in range(n)]\n",
    "s = ['+' for i in range(n)]\n",
    "g = 'g2'\n",
    "e = [[x + 200 for x in e1] for e1 in e]\n",
    "nt = 'g2[1,1,3]'\n",
    "t = 'g2_t4'\n",
    "ag1 = 'novel'\n",
    "ag2 = 'p1'\n",
    "df = make_exon_df(n,c,e,s,g,t,nt,ag1,ag2)\n",
    "df['tss_id'] = 'g2_1'\n",
    "df['tes_id'] = 'g2_3'\n",
    "ts.append(df)\n",
    "\n",
    "# t9, g2 monoexonic <80% overlap\n",
    "e = [[50,150]]\n",
    "n = len(e)\n",
    "c = ['chr1' for i in range(n)]\n",
    "s = ['+' for i in range(n)]\n",
    "g = 'g2'\n",
    "e = [[x + 200 for x in e1] for e1 in e]\n",
    "nt = 'g2[1,1,3]'\n",
    "t = 'g2_t4'\n",
    "ag1 = 'novel'\n",
    "ag2 = 'p1'\n",
    "df = make_exon_df(n,c,e,s,g,t,nt,ag1,ag2)\n",
    "df['tss_id'] = 'g2_1'\n",
    "df['tes_id'] = 'g2_3'\n",
    "ts.append(df)\n",
    "\n",
    "# t6n-- 5/3' ends longer/ shorter, same IC\n",
    "e = [[1,30], [40,50]]\n",
    "n = len(e)\n",
    "c = ['chr1' for i in range(n)]\n",
    "s = ['+' for i in range(n)]\n",
    "g = 'g3'\n",
    "e = [[x + 300 for x in e1] for e1 in e]\n",
    "nt = 'g3[2,1,2]'\n",
    "t = 'g3_t6'\n",
    "ag1 = 'known'\n",
    "ag2 = 'p3'\n",
    "df = make_exon_df(n,c,e,s,g,t,nt,ag1,ag2)\n",
    "df['tss_id'] = 'g3_1'\n",
    "df['tes_id'] = 'g3_1'\n",
    "ts.append(df)\n",
    "\n",
    "# t8\n",
    "e = [[10,25], [35,45]]\n",
    "n = len(e)\n",
    "c = ['chr1' for i in range(n)]\n",
    "s = ['+' for i in range(n)]\n",
    "g = 'g4'\n",
    "e = [[x + 400 for x in e1] for e1 in e]\n",
    "nt = 'g4[1,2,1]'\n",
    "t = 'g4_t8'\n",
    "ag1 = 'known'\n",
    "ag2 = 'p3'\n",
    "df = make_exon_df(n,c,e,s,g,t,nt,ag1,ag2)\n",
    "df['tss_id'] = 'g4_1'\n",
    "df['tes_id'] = 'g4_1'\n",
    "ts.append(df)\n",
    "\n",
    "# make test gtf\n",
    "test_df = make_test_gtf(ts)\n",
    "test_df.rename({'transcript_id': 'original_transcript_id',\n",
    "                'transcript_name': 'original_transcript_name'},\n",
    "               axis=1, inplace=True)\n",
    "test_df.rename({'new_transcript_id': 'transcript_id'},\n",
    "               axis=1, inplace=True)\n",
    "test_df['transcript_name'] = test_df['transcript_id']\n",
    "test_df = pr.PyRanges(test_df)\n",
    "test_df.to_gtf('test_gtf_2.gtf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "5c6c0e05-84d3-477d-b256-b57e570fac5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame()\n",
    "for gtf in ['test_gtf_1.gtf', 'test_gtf_2.gtf']:\n",
    "    temp = pr.read_gtf(gtf).df\n",
    "    df = pd.concat([df, temp], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "69b9a941-530f-401b-b2df-6bb04e0685ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pr.PyRanges(df)\n",
    "df.to_gtf('test_gtf.gtf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3236ddf-ce58-4ebf-8e31-821a1bfa6f72",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf66ccbb-554c-4f1d-9ee3-7dfe028c4c82",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6f2753d-f27a-4da8-8550-b7cbdb677b4e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b12b1352-0609-48ac-a7f0-f36705af9f5e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31a8df4d-aade-4a48-b6a4-c722e67091b6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0528b82-610e-4ea7-b392-3346ff08ea23",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9fd94f9-9740-4cb0-ae3a-1aba0b23892f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1a184d4-2906-4ba1-bd98-44dbf8c9b781",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
