{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d847b398-99fe-46b1-a9de-0dc245c2e891",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import pyranges as pr\n",
    "import cerberus\n",
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1790625c-214b-4655-9e04-b6f7ea016748",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_hier_entry(df, how='t'):\n",
    "    \"\"\"\n",
    "    kind {'g','t'}\n",
    "    \"\"\"\n",
    "    agg_dict = {'min_coord': 'min', 'max_coord': 'max'}\n",
    "    t_df = df.copy(deep=True)\n",
    "    t_df['min_coord'] = t_df[['Start', 'End']].min(axis=1)\n",
    "    t_df['max_coord'] = t_df[['Start', 'End']].max(axis=1)\n",
    "    if how == 't':\n",
    "        gb_cols = ['Chromosome', 'Strand', 'gene_name',\n",
    "                   'gene_id', 'transcript_id', 'transcript_name',\n",
    "                   'tss_id', 'tes_id',\n",
    "                   'new_transcript_id', 'original_transcript_id',\n",
    "                   'original_transcript_name', 'ag1', 'ag2']\n",
    "    elif how == 'g':\n",
    "        gb_cols = ['Chromosome', 'Strand', 'gene_name',\n",
    "                   'gene_id']\n",
    "    gb_cols = list(set(gb_cols)&(set(t_df.columns)))\n",
    "\n",
    "    cols = gb_cols + ['min_coord', 'max_coord']\n",
    "    t_df = t_df[cols]\n",
    "    t_df = t_df.groupby(gb_cols, observed=True).agg(agg_dict).reset_index()\n",
    "    t_df.rename({'min_coord': 'Start', 'max_coord': 'End'}, axis=1, inplace=True)\n",
    "    if how == 't':\n",
    "        t_df['Feature'] = 'transcript'\n",
    "    elif how == 'g':\n",
    "        t_df['Feature'] = 'gene'\n",
    "\n",
    "    return t_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "a3c9ecd4-4ee5-4122-8d85-1e4d65b39341",
   "metadata": {},
   "outputs": [],
   "source": [
    "ifile = 'HG00621_1_novel_gene_rename.gtf'\n",
    "ref_file = '../../../ref/ref.gtf'\n",
    "ofile = 'test.gtf'\n",
    "tool = 'flair'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "d41fb56b-0d63-4375-89f4-b2d8d1b7fe6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pr.read_gtf(ifile).df\n",
    "\n",
    "# limit to just transcripts and exons b/c we'll just\n",
    "# reconstruct the genes\n",
    "df = df.loc[df.Feature.isin(['transcript', 'exon'])]\n",
    "gtf_df = df.copy(deep=True)\n",
    "\n",
    "# flair -- rename antisense genes\n",
    "if tool == 'flair':\n",
    "    ref_df = pr.read_gtf(ref_file).df\n",
    "\n",
    "    # get strand, gene id, and transcript id of flair transcripts\n",
    "    df = df[['gene_id', 'transcript_id', 'Strand']].drop_duplicates()\n",
    "    df['gid_stable'] = cerberus.get_stable_gid(df, 'gene_id')\n",
    "\n",
    "    # get strand and gene id for reference\n",
    "    ref_df = ref_df[['gene_id', 'Strand']].drop_duplicates()\n",
    "    ref_df['gid_stable'] = cerberus.get_stable_gid(ref_df, 'gene_id')\n",
    "\n",
    "    # merge w/ reference to figure out which transcripts are antisense\n",
    "    df = df.merge(ref_df,\n",
    "                  how='left',\n",
    "                  on='gid_stable',\n",
    "                  suffixes=('', '_ref'))\n",
    "\n",
    "    # limit to things on the opposite strand as ref\n",
    "    # and known things\n",
    "    # and replace gene id for those\n",
    "    temp = df.loc[(df.Strand!=df.Strand_ref)&\\\n",
    "                  (df.Strand_ref.notnull())]\n",
    "    tids = temp.transcript_id.unique().tolist()\n",
    "    inds = gtf_df.loc[gtf_df.transcript_id.isin(tids)].index\n",
    "    gtf_df.loc[inds, 'gene_id'] = gtf_df.loc[inds, 'gene_id']+'_antisense'\n",
    "\n",
    "    l1 = len(gtf_df.gene_id.unique())\n",
    "    l2 = len(gtf_df[['Strand', 'gene_id']].drop_duplicates())\n",
    "\n",
    "    # make sure we have the same number of gene ids at the end (l2)\n",
    "    # as we did unique gene id + strand combos at the beginning (l1)\n",
    "    assert l1 == l2\n",
    "\n",
    "    # definitionally we are making novel loci by merging \n",
    "    # on strands. So make sure we don't have any novel genes\n",
    "    # w/ antisense designation\n",
    "    assert len(df.loc[(df.gene_id.str.contains('LOC'))&\\\n",
    "                 (df.gene_id.str.contains('antisense'))].index) == 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "4bcaa224-fc50-4ddd-a587-271ac009b9ac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3961\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "print(len(df.loc[df.gene_id.str.contains('LOC')].index))\n",
    "print(len(df.loc[(df.gene_id.str.contains('LOC'))&\\\n",
    "                 (df.gene_id.str.contains('antisense'))].index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "cb7111dd-4c54-4f4f-99b1-3a5870fef715",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make new gene entries\n",
    "df = gtf_df.copy(deep=True)\n",
    "l1 = len(df.gene_id.unique().tolist())\n",
    "\n",
    "# make gene entry\n",
    "g_df = make_hier_entry(df, how='g')\n",
    "\n",
    "if tool == 'espresso': source = 'Espresso'\n",
    "elif tool == 'flair': source = 'FLAIR'\n",
    "elif tool == 'iq': source = 'IsoQuant'\n",
    "g_df['Source'] = source\n",
    "g_df['Frame'] = '.'\n",
    "g_df['Score'] = '.'\n",
    "l2 = len(g_df.loc[g_df.Feature=='gene'].index)\n",
    "assert l1 == l2\n",
    "\n",
    "# concat them and then sort gtf\n",
    "df = pd.concat([df, g_df], axis=0)\n",
    "df = cerberus.sort_gtf(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "1f4a726a-b72a-41a8-976f-cee4e78d0048",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>gene</td>\n",
       "      <td>169795042</td>\n",
       "      <td>169804386</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>transcript</td>\n",
       "      <td>169795078</td>\n",
       "      <td>169804386</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>98e251b6-26ca-48f1-a06f-0163f0ad2ee0:0-0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>exon</td>\n",
       "      <td>169795078</td>\n",
       "      <td>169795213</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>98e251b6-26ca-48f1-a06f-0163f0ad2ee0:0-0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>exon</td>\n",
       "      <td>169798918</td>\n",
       "      <td>169798958</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>98e251b6-26ca-48f1-a06f-0163f0ad2ee0:0-0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FLAIR</td>\n",
       "      <td>exon</td>\n",
       "      <td>169800882</td>\n",
       "      <td>169800971</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>98e251b6-26ca-48f1-a06f-0163f0ad2ee0:0-0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome Source     Feature      Start        End Score Strand Frame  \\\n",
       "0       chr1  FLAIR        gene  169795042  169804386     .      +     .   \n",
       "1       chr1  FLAIR  transcript  169795078  169804386     .      +     .   \n",
       "2       chr1  FLAIR        exon  169795078  169795213     .      +     .   \n",
       "3       chr1  FLAIR        exon  169798918  169798958     .      +     .   \n",
       "4       chr1  FLAIR        exon  169800882  169800971     .      +     .   \n",
       "\n",
       "              gene_id                             transcript_id exon_number  \n",
       "0  ENSG00000000460.17                                       NaN         NaN  \n",
       "1  ENSG00000000460.17  98e251b6-26ca-48f1-a06f-0163f0ad2ee0:0-0         NaN  \n",
       "2  ENSG00000000460.17  98e251b6-26ca-48f1-a06f-0163f0ad2ee0:0-0           0  \n",
       "3  ENSG00000000460.17  98e251b6-26ca-48f1-a06f-0163f0ad2ee0:0-0           1  \n",
       "4  ENSG00000000460.17  98e251b6-26ca-48f1-a06f-0163f0ad2ee0:0-0           2  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4b0abd6a-a34c-4471-84e9-38cd8c7be6af",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11965\n",
      "11965\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "ec438480-1b50-46ba-81df-2ecb93ae32bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # iq -- add gene names and transcript names\n",
    "# if tool == 'iq':\n",
    "\n",
    "#     # just grab all the gene ids to be the gene names\n",
    "#     df['gene_name'] = df['gene_id']\n",
    "\n",
    "#     # for transcripts / other transcript-level features do the\n",
    "#     # same but restrict to those feats\n",
    "#     inds = df.loc[df.Feature != 'gene'].index\n",
    "#     df.loc[inds, 'transcript_name'] = df.loc[inds, 'transcript_id']\n",
    "\n",
    "# convert + save\n",
    "df = pr.PyRanges(df)\n",
    "df.to_gtf(ofile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34f6bb04-a33f-454e-9a16-0895b0ec72c0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
