{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f9499390-7181-46d6-944b-1785b4ca025d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import  cerberus\n",
    "import glob\n",
    "import pyranges as pr\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "dadd8446-4d42-45c7-b390-7c2f17485859",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GM10493_1.gtf\n",
      "GM12878_1.gtf\n"
     ]
    }
   ],
   "source": [
    "df = pd.DataFrame()\n",
    "for f in glob.glob('*gtf'):\n",
    "    print(f)\n",
    "    df = pd.concat([pr.read_gtf(f, as_df=True), df], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3ac15b43-7cfa-4e66-a339-47fd19560a0c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>140848</th>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>ENSG00000000419[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133260</th>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>ENSG00000000419[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140868</th>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>ENSG00000000419[1,14,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133280</th>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>ENSG00000000419[1,14,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133270</th>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>ENSG00000000419[1,16,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200267</th>\n",
       "      <td>ENSG00000310458.1</td>\n",
       "      <td>ENSG00000310458[1,3,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209601</th>\n",
       "      <td>ENSG00000310521.1</td>\n",
       "      <td>ENSG00000310521[2,10,3]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200696</th>\n",
       "      <td>ENSG00000310521.1</td>\n",
       "      <td>ENSG00000310521[2,10,3]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84178</th>\n",
       "      <td>ENSG00000310525.1</td>\n",
       "      <td>ENSG00000310525[2,3,2]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89099</th>\n",
       "      <td>ENSG00000310525.1</td>\n",
       "      <td>ENSG00000310525[2,3,2]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29534 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   gene_id            transcript_id\n",
       "140848  ENSG00000000419.14   ENSG00000000419[1,1,1]\n",
       "133260  ENSG00000000419.14   ENSG00000000419[1,1,1]\n",
       "140868  ENSG00000000419.14  ENSG00000000419[1,14,1]\n",
       "133280  ENSG00000000419.14  ENSG00000000419[1,14,1]\n",
       "133270  ENSG00000000419.14  ENSG00000000419[1,16,1]\n",
       "...                    ...                      ...\n",
       "200267   ENSG00000310458.1   ENSG00000310458[1,3,1]\n",
       "209601   ENSG00000310521.1  ENSG00000310521[2,10,3]\n",
       "200696   ENSG00000310521.1  ENSG00000310521[2,10,3]\n",
       "84178    ENSG00000310525.1   ENSG00000310525[2,3,2]\n",
       "89099    ENSG00000310525.1   ENSG00000310525[2,3,2]\n",
       "\n",
       "[29534 rows x 2 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = df.loc[df.Feature=='transcript']\n",
    "temp.loc[temp.transcript_id.duplicated(keep=False)].sort_values(by='transcript_id')[['gene_id', 'transcript_id']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7705e669-43b7-4f06-b745-d2e0f1220375",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "12606"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df.loc[df.gene_id.str.contains('novel_gene')].index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7093665f-0893-4f2b-8200-7cea547bd9a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df.loc[(df.gene_id.str.contains('novel_gene'))&\\\n",
    "           (df.transcript_id.isnull())&\\\n",
    "           (df.Feature=='transcript')].index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "40e33950-facb-48f4-a59f-42b481159f02",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df.loc[df.gene_id.str.contains('novel_gene'), 'Chromosome'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0c95e0a1-bc96-4576-b7e8-d3f5615d4994",
   "metadata": {},
   "outputs": [],
   "source": [
    "# all of these problems are from novel genes, \n",
    "# and all novel genes are problems\n",
    "\n",
    "# the novel bois didnt get added to h5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d81b315-4801-43ef-bbc7-634034d0a2ed",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2ae54b0c-d2db-4aef-9094-6d25f56804a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>exons</th>\n",
       "      <th>tss_first_sd_issue</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>tes_last_sa_issue</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>hgnc_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>140848</th>\n",
       "      <td>chr20</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50934856</td>\n",
       "      <td>50958580</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000371588.10</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSP00000360644.5</td>\n",
       "      <td>OTTHUMT00000079716.1</td>\n",
       "      <td>CCDS13434.1</td>\n",
       "      <td>OTTHUMG00000032742.2</td>\n",
       "      <td>HGNC:3005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133260</th>\n",
       "      <td>chr20</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50934856</td>\n",
       "      <td>50958580</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000371588.10</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSP00000360644.5</td>\n",
       "      <td>OTTHUMT00000079716.1</td>\n",
       "      <td>CCDS13434.1</td>\n",
       "      <td>OTTHUMG00000032742.2</td>\n",
       "      <td>HGNC:3005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140868</th>\n",
       "      <td>chr20</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50934856</td>\n",
       "      <td>50958580</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000683048.1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000506986.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMG00000032742.2</td>\n",
       "      <td>HGNC:3005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133280</th>\n",
       "      <td>chr20</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50934856</td>\n",
       "      <td>50958580</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000683048.1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000506986.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMG00000032742.2</td>\n",
       "      <td>HGNC:3005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133270</th>\n",
       "      <td>chr20</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50934856</td>\n",
       "      <td>50958580</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000466152.5</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSP00000507119.1</td>\n",
       "      <td>OTTHUMT00000079717.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMG00000032742.2</td>\n",
       "      <td>HGNC:3005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200267</th>\n",
       "      <td>chr7</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>44194249</td>\n",
       "      <td>44198255</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000310458.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000849948.1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209601</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>119855428</td>\n",
       "      <td>119869023</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000310521.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000765898.1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMG00000195600.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200696</th>\n",
       "      <td>chr8</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>119855428</td>\n",
       "      <td>119869023</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000310521.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000765898.1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTHUMG00000195600.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84178</th>\n",
       "      <td>chr16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript</td>\n",
       "      <td>72084845</td>\n",
       "      <td>72093681</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000310525.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89099</th>\n",
       "      <td>chr16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript</td>\n",
       "      <td>72084845</td>\n",
       "      <td>72093681</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000310525.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29534 rows Ã— 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome  Source     Feature      Start        End Score Strand  \\\n",
       "140848      chr20  HAVANA  transcript   50934856   50958580     .      -   \n",
       "133260      chr20  HAVANA  transcript   50934856   50958580     .      -   \n",
       "140868      chr20  HAVANA  transcript   50934856   50958580     .      -   \n",
       "133280      chr20  HAVANA  transcript   50934856   50958580     .      -   \n",
       "133270      chr20  HAVANA  transcript   50934856   50958580     .      -   \n",
       "...           ...     ...         ...        ...        ...   ...    ...   \n",
       "200267       chr7  HAVANA  transcript   44194249   44198255     .      +   \n",
       "209601       chr8  HAVANA  transcript  119855428  119869023     .      +   \n",
       "200696       chr8  HAVANA  transcript  119855428  119869023     .      +   \n",
       "84178       chr16     NaN  transcript   72084845   72093681     .      -   \n",
       "89099       chr16     NaN  transcript   72084845   72093681     .      -   \n",
       "\n",
       "       Frame             gene_id       gene_type  ... exons  \\\n",
       "140848     .  ENSG00000000419.14  protein_coding  ...     9   \n",
       "133260     .  ENSG00000000419.14  protein_coding  ...     9   \n",
       "140868     .  ENSG00000000419.14  protein_coding  ...     7   \n",
       "133280     .  ENSG00000000419.14  protein_coding  ...     7   \n",
       "133270     .  ENSG00000000419.14  protein_coding  ...     9   \n",
       "...      ...                 ...             ...  ...   ...   \n",
       "200267     .   ENSG00000310458.1          lncRNA  ...     2   \n",
       "209601     .   ENSG00000310521.1          lncRNA  ...     2   \n",
       "200696     .   ENSG00000310521.1          lncRNA  ...     2   \n",
       "84178      .   ENSG00000310525.1             NaN  ...   NaN   \n",
       "89099      .   ENSG00000310525.1             NaN  ...   NaN   \n",
       "\n",
       "       tss_first_sd_issue original_transcript_id tes_last_sa_issue  \\\n",
       "140848              False     ENST00000371588.10             False   \n",
       "133260              False     ENST00000371588.10             False   \n",
       "140868              False      ENST00000683048.1             False   \n",
       "133280              False      ENST00000683048.1             False   \n",
       "133270              False      ENST00000466152.5             False   \n",
       "...                   ...                    ...               ...   \n",
       "200267              False      ENST00000849948.1             False   \n",
       "209601              False      ENST00000765898.1             False   \n",
       "200696              False      ENST00000765898.1             False   \n",
       "84178                 NaN                    NaN               NaN   \n",
       "89099                 NaN                    NaN               NaN   \n",
       "\n",
       "       transcript_support_level         protein_id     havana_transcript  \\\n",
       "140848                        1  ENSP00000360644.5  OTTHUMT00000079716.1   \n",
       "133260                        1  ENSP00000360644.5  OTTHUMT00000079716.1   \n",
       "140868                      NaN  ENSP00000506986.1                   NaN   \n",
       "133280                      NaN  ENSP00000506986.1                   NaN   \n",
       "133270                        1  ENSP00000507119.1  OTTHUMT00000079717.1   \n",
       "...                         ...                ...                   ...   \n",
       "200267                      NaN                NaN                   NaN   \n",
       "209601                      NaN                NaN                   NaN   \n",
       "200696                      NaN                NaN                   NaN   \n",
       "84178                       NaN                NaN                   NaN   \n",
       "89099                       NaN                NaN                   NaN   \n",
       "\n",
       "             ccdsid           havana_gene    hgnc_id  \n",
       "140848  CCDS13434.1  OTTHUMG00000032742.2  HGNC:3005  \n",
       "133260  CCDS13434.1  OTTHUMG00000032742.2  HGNC:3005  \n",
       "140868          NaN  OTTHUMG00000032742.2  HGNC:3005  \n",
       "133280          NaN  OTTHUMG00000032742.2  HGNC:3005  \n",
       "133270          NaN  OTTHUMG00000032742.2  HGNC:3005  \n",
       "...             ...                   ...        ...  \n",
       "200267          NaN                   NaN        NaN  \n",
       "209601          NaN  OTTHUMG00000195600.1        NaN  \n",
       "200696          NaN  OTTHUMG00000195600.1        NaN  \n",
       "84178           NaN                   NaN        NaN  \n",
       "89099           NaN                   NaN        NaN  \n",
       "\n",
       "[29534 rows x 32 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = df.loc[df.Feature=='transcript']\n",
    "temp.loc[temp.transcript_id.duplicated(keep=False)].sort_values(by='transcript_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "dfdf012e-cae7-4acd-8976-d3fe59090cd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = cerberus.sort_gtf(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "fea40b3c-cbca-4f2a-bc84-71291659b84e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>source</th>\n",
       "      <th>novelty</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tss</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GL000008.2</td>\n",
       "      <td>83306</td>\n",
       "      <td>83446</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000296732_1</td>\n",
       "      <td>47,GM12878_1</td>\n",
       "      <td>Known</td>\n",
       "      <td>ENSG00000296732</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GL000008.2</td>\n",
       "      <td>85532</td>\n",
       "      <td>85583</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000296732_2</td>\n",
       "      <td>47</td>\n",
       "      <td>Known</td>\n",
       "      <td>ENSG00000296732</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GL000008.2</td>\n",
       "      <td>208519</td>\n",
       "      <td>208637</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000296756_1</td>\n",
       "      <td>47</td>\n",
       "      <td>Known</td>\n",
       "      <td>ENSG00000296756</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GL000008.2</td>\n",
       "      <td>170299</td>\n",
       "      <td>170454</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000296775_1</td>\n",
       "      <td>47,GM10493_1,GM12878_1</td>\n",
       "      <td>Known</td>\n",
       "      <td>ENSG00000296775</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GL000009.2</td>\n",
       "      <td>35368</td>\n",
       "      <td>35419</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000296902_1</td>\n",
       "      <td>47</td>\n",
       "      <td>Known</td>\n",
       "      <td>ENSG00000296902</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GL000009.2</td>\n",
       "      <td>5095</td>\n",
       "      <td>5222</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000297619_1</td>\n",
       "      <td>47</td>\n",
       "      <td>Known</td>\n",
       "      <td>ENSG00000297619</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GL000009.2</td>\n",
       "      <td>82811</td>\n",
       "      <td>82862</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000306744_1</td>\n",
       "      <td>47</td>\n",
       "      <td>Known</td>\n",
       "      <td>ENSG00000306744</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GL000009.2</td>\n",
       "      <td>43268</td>\n",
       "      <td>43322</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000306767_1</td>\n",
       "      <td>47</td>\n",
       "      <td>Known</td>\n",
       "      <td>ENSG00000306767</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------+\n",
       "| Chromosome   |     Start |       End | Strand       | +5    |\n",
       "| (category)   |   (int64) |   (int64) | (category)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+-------|\n",
       "| GL000008.2   |     83306 |     83446 | +            | ...   |\n",
       "| GL000008.2   |     85532 |     85583 | +            | ...   |\n",
       "| GL000008.2   |    208519 |    208637 | -            | ...   |\n",
       "| GL000008.2   |    170299 |    170454 | -            | ...   |\n",
       "| GL000009.2   |     35368 |     35419 | +            | ...   |\n",
       "| GL000009.2   |      5095 |      5222 | +            | ...   |\n",
       "| GL000009.2   |     82811 |     82862 | +            | ...   |\n",
       "| GL000009.2   |     43268 |     43322 | +            | ...   |\n",
       "+--------------+-----------+-----------+--------------+-------+\n",
       "Stranded PyRanges object has 8 rows and 9 columns from 2 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "5 hidden columns: tss_id, source, novelty, gene_id, tss"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ca = cerberus.read('../ref/ref.h5')\n",
    "\n",
    "tss = ca.tss\n",
    "tss.rename({'Name': 'tss_id'}, axis=1, inplace=True)\n",
    "tes = ca.tes\n",
    "tes.rename({'Name': 'tes_id'}, axis=1, inplace=True)\n",
    "\n",
    "tss = pr.PyRanges(tss)\n",
    "tes = pr.PyRanges(tes)\n",
    "\n",
    "df = cerberus.update_gtf_ends(df, tss, tes)\n",
    "tss.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "9433cae4-60c6-4ded-b286-2b3f3e26cc05",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>exons</th>\n",
       "      <th>tss_first_sd_issue</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>tes_last_sa_issue</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>hgnc_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gene</td>\n",
       "      <td>169794786</td>\n",
       "      <td>169853095</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gene</td>\n",
       "      <td>169794786</td>\n",
       "      <td>169853095</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript</td>\n",
       "      <td>169795014</td>\n",
       "      <td>169804396</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon</td>\n",
       "      <td>169795014</td>\n",
       "      <td>169795213</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon</td>\n",
       "      <td>169798917</td>\n",
       "      <td>169798958</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome Source     Feature      Start        End Score Strand Frame  \\\n",
       "0       chr1    NaN        gene  169794786  169853095     .      +     .   \n",
       "1       chr1    NaN        gene  169794786  169853095     .      +     .   \n",
       "2       chr1    NaN  transcript  169795014  169804396     .      +     .   \n",
       "3       chr1    NaN        exon  169795014  169795213     .      +     .   \n",
       "4       chr1    NaN        exon  169798917  169798958     .      +     .   \n",
       "\n",
       "              gene_id       gene_type  ... exons tss_first_sd_issue  \\\n",
       "0  ENSG00000000460.17  protein_coding  ...   NaN                NaN   \n",
       "1  ENSG00000000460.17  protein_coding  ...   NaN                NaN   \n",
       "2  ENSG00000000460.17             NaN  ...   NaN                NaN   \n",
       "3  ENSG00000000460.17             NaN  ...   NaN                NaN   \n",
       "4  ENSG00000000460.17             NaN  ...   NaN                NaN   \n",
       "\n",
       "  original_transcript_id tes_last_sa_issue transcript_support_level  \\\n",
       "0                    NaN               NaN                      NaN   \n",
       "1                    NaN               NaN                      NaN   \n",
       "2                    NaN               NaN                      NaN   \n",
       "3                    NaN               NaN                      NaN   \n",
       "4                    NaN               NaN                      NaN   \n",
       "\n",
       "  protein_id havana_transcript ccdsid havana_gene hgnc_id  \n",
       "0        NaN               NaN    NaN         NaN     NaN  \n",
       "1        NaN               NaN    NaN         NaN     NaN  \n",
       "2        NaN               NaN    NaN         NaN     NaN  \n",
       "3        NaN               NaN    NaN         NaN     NaN  \n",
       "4        NaN               NaN    NaN         NaN     NaN  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7e9a1c68-7488-43ad-8208-4f0422455c65",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "447907\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/Documents/programming/mortazavi_lab/bin/cerberus/cerberus/cerberus.py:1184: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  t_df[non_gb_cols].fillna('', inplace=True)\n",
      "/Users/fairliereese/Documents/programming/mortazavi_lab/bin/cerberus/cerberus/cerberus.py:1195: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df.drop(non_gb_cols, axis=1, inplace=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "447892\n"
     ]
    }
   ],
   "source": [
    "print(len(df.index))\n",
    "df = cerberus.agg_gtf(df)\n",
    "print(len(df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "29f9f629-b357-4203-b6b0-f416598cfe3c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>tss_first_sd_issue</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>tes_last_sa_issue</th>\n",
       "      <th>hgnc_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>221622</th>\n",
       "      <td>chr20</td>\n",
       "      <td>gene</td>\n",
       "      <td>50934856</td>\n",
       "      <td>50958580</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221621</th>\n",
       "      <td>chr20</td>\n",
       "      <td>gene</td>\n",
       "      <td>50934856</td>\n",
       "      <td>50958580</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221724</th>\n",
       "      <td>chr1</td>\n",
       "      <td>gene</td>\n",
       "      <td>169849620</td>\n",
       "      <td>169893984</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000457.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000000457.14</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221723</th>\n",
       "      <td>chr1</td>\n",
       "      <td>gene</td>\n",
       "      <td>169849620</td>\n",
       "      <td>169893984</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000457.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000000457.14</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>gene</td>\n",
       "      <td>169794786</td>\n",
       "      <td>169853095</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>447130</th>\n",
       "      <td>chr7</td>\n",
       "      <td>gene</td>\n",
       "      <td>716607</td>\n",
       "      <td>727719</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>novel_gene_chr7_59</td>\n",
       "      <td>NaN</td>\n",
       "      <td>novel_gene_chr7_59</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>447362</th>\n",
       "      <td>chr9</td>\n",
       "      <td>gene</td>\n",
       "      <td>37400497</td>\n",
       "      <td>37402888</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>novel_gene_chr9_1367</td>\n",
       "      <td>NaN</td>\n",
       "      <td>novel_gene_chr9_1367</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221110</th>\n",
       "      <td>chr9</td>\n",
       "      <td>gene</td>\n",
       "      <td>37407261</td>\n",
       "      <td>37410166</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>novel_gene_chr9_1367</td>\n",
       "      <td>NaN</td>\n",
       "      <td>novel_gene_chr9_1367</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221227</th>\n",
       "      <td>chr9</td>\n",
       "      <td>gene</td>\n",
       "      <td>128618891</td>\n",
       "      <td>128624727</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>novel_gene_chr9_3435</td>\n",
       "      <td>NaN</td>\n",
       "      <td>novel_gene_chr9_3435</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>447517</th>\n",
       "      <td>chr9</td>\n",
       "      <td>gene</td>\n",
       "      <td>128160255</td>\n",
       "      <td>128163949</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>novel_gene_chr9_3435</td>\n",
       "      <td>NaN</td>\n",
       "      <td>novel_gene_chr9_3435</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18158 rows Ã— 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome Feature      Start        End Score Strand Frame  \\\n",
       "221622      chr20    gene   50934856   50958580     .      -     .   \n",
       "221621      chr20    gene   50934856   50958580     .      -     .   \n",
       "221724       chr1    gene  169849620  169893984     .      -     .   \n",
       "221723       chr1    gene  169849620  169893984     .      -     .   \n",
       "0            chr1    gene  169794786  169853095     .      +     .   \n",
       "...           ...     ...        ...        ...   ...    ...   ...   \n",
       "447130       chr7    gene     716607     727719     .      -     .   \n",
       "447362       chr9    gene   37400497   37402888     .      -     .   \n",
       "221110       chr9    gene   37407261   37410166     .      +     .   \n",
       "221227       chr9    gene  128618891  128624727     .      +     .   \n",
       "447517       chr9    gene  128160255  128163949     .      -     .   \n",
       "\n",
       "                     gene_id       gene_type             gene_name  ...  \\\n",
       "221622    ENSG00000000419.14  protein_coding    ENSG00000000419.14  ...   \n",
       "221621    ENSG00000000419.14  protein_coding    ENSG00000000419.14  ...   \n",
       "221724    ENSG00000000457.14  protein_coding    ENSG00000000457.14  ...   \n",
       "221723    ENSG00000000457.14  protein_coding    ENSG00000000457.14  ...   \n",
       "0         ENSG00000000460.17  protein_coding    ENSG00000000460.17  ...   \n",
       "...                      ...             ...                   ...  ...   \n",
       "447130    novel_gene_chr7_59             NaN    novel_gene_chr7_59  ...   \n",
       "447362  novel_gene_chr9_1367             NaN  novel_gene_chr9_1367  ...   \n",
       "221110  novel_gene_chr9_1367             NaN  novel_gene_chr9_1367  ...   \n",
       "221227  novel_gene_chr9_3435             NaN  novel_gene_chr9_3435  ...   \n",
       "447517  novel_gene_chr9_3435             NaN  novel_gene_chr9_3435  ...   \n",
       "\n",
       "       tss_first_sd_issue original_transcript_name transcript_support_level  \\\n",
       "221622                NaN                      NaN                      NaN   \n",
       "221621                NaN                      NaN                      NaN   \n",
       "221724                NaN                      NaN                      NaN   \n",
       "221723                NaN                      NaN                      NaN   \n",
       "0                     NaN                      NaN                      NaN   \n",
       "...                   ...                      ...                      ...   \n",
       "447130                NaN                      NaN                      NaN   \n",
       "447362                NaN                      NaN                      NaN   \n",
       "221110                NaN                      NaN                      NaN   \n",
       "221227                NaN                      NaN                      NaN   \n",
       "447517                NaN                      NaN                      NaN   \n",
       "\n",
       "       ccdsid havana_transcript  tag havana_gene protein_id tes_last_sa_issue  \\\n",
       "221622    NaN               NaN  NaN         NaN        NaN               NaN   \n",
       "221621    NaN               NaN  NaN         NaN        NaN               NaN   \n",
       "221724    NaN               NaN  NaN         NaN        NaN               NaN   \n",
       "221723    NaN               NaN  NaN         NaN        NaN               NaN   \n",
       "0         NaN               NaN  NaN         NaN        NaN               NaN   \n",
       "...       ...               ...  ...         ...        ...               ...   \n",
       "447130    NaN               NaN  NaN         NaN        NaN               NaN   \n",
       "447362    NaN               NaN  NaN         NaN        NaN               NaN   \n",
       "221110    NaN               NaN  NaN         NaN        NaN               NaN   \n",
       "221227    NaN               NaN  NaN         NaN        NaN               NaN   \n",
       "447517    NaN               NaN  NaN         NaN        NaN               NaN   \n",
       "\n",
       "       hgnc_id  \n",
       "221622     NaN  \n",
       "221621     NaN  \n",
       "221724     NaN  \n",
       "221723     NaN  \n",
       "0          NaN  \n",
       "...        ...  \n",
       "447130     NaN  \n",
       "447362     NaN  \n",
       "221110     NaN  \n",
       "221227     NaN  \n",
       "447517     NaN  \n",
       "\n",
       "[18158 rows x 32 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = df.loc[df.Feature=='gene']\n",
    "temp.loc[temp.gene_id.duplicated(keep=False)].sort_values(by='gene_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "5f805394-10fc-400b-86ee-fa5c0fec4840",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Score</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tes_id</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>tss</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>tes</th>\n",
       "      <th>Frame</th>\n",
       "      <th>Start</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>ic_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ic</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>221622</th>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50958580</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>50934856</td>\n",
       "      <td>chr20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221621</th>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50958580</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>50934856</td>\n",
       "      <td>chr20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221724</th>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSG00000000457.14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>169893984</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>169849620</td>\n",
       "      <td>chr1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000000457.14</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221723</th>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSG00000000457.14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>169893984</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>169849620</td>\n",
       "      <td>chr1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000000457.14</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>169853095</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>169794786</td>\n",
       "      <td>chr1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       transcript_id Score       gene_type Feature             gene_id tes_id  \\\n",
       "221622           NaN     .  protein_coding    gene  ENSG00000000419.14    NaN   \n",
       "221621           NaN     .  protein_coding    gene  ENSG00000000419.14    NaN   \n",
       "221724           NaN     .  protein_coding    gene  ENSG00000000457.14    NaN   \n",
       "221723           NaN     .  protein_coding    gene  ENSG00000000457.14    NaN   \n",
       "0                NaN     .  protein_coding    gene  ENSG00000000460.17    NaN   \n",
       "\n",
       "       tss_id  tss        End Strand transcript_name  tes Frame      Start  \\\n",
       "221622    NaN  NaN   50958580      -             NaN  NaN     .   50934856   \n",
       "221621    NaN  NaN   50958580      -             NaN  NaN     .   50934856   \n",
       "221724    NaN  NaN  169893984      -             NaN  NaN     .  169849620   \n",
       "221723    NaN  NaN  169893984      -             NaN  NaN     .  169849620   \n",
       "0         NaN  NaN  169853095      +             NaN  NaN     .  169794786   \n",
       "\n",
       "       Chromosome ic_id           gene_name   ic  \n",
       "221622      chr20   NaN  ENSG00000000419.14  NaN  \n",
       "221621      chr20   NaN  ENSG00000000419.14  NaN  \n",
       "221724       chr1   NaN  ENSG00000000457.14  NaN  \n",
       "221723       chr1   NaN  ENSG00000000457.14  NaN  \n",
       "0            chr1   NaN  ENSG00000000460.17  NaN  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# duplicated genes?\n",
    "gb_cols = ['Chromosome',\n",
    "             'Feature',\n",
    "             'Start', 'End',\n",
    "             'Score', 'Strand', 'Frame', 'gene_id', 'gene_name',\n",
    "             'gene_status', 'gene_type', 'talon_gene',\n",
    "             'ic', 'ic_id', 'tss_id', 'tss', 'tes_id', 'tes', 'transcript_id',\n",
    "             'transcript_name']\n",
    "gb_cols = list(set(df.columns)&set(gb_cols))\n",
    "temp = df.loc[df.Feature=='gene']\n",
    "temp.loc[temp.gene_id.duplicated(keep=False)].sort_values(by='gene_id')[gb_cols].head()\n",
    "# assert len(temp.loc[temp.gene_id.duplicated()].index) == 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb514d78-e7d2-4083-a892-048e1d02b092",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "95e4218e-b9b0-43f6-9b32-9c3d4ac04a3b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Score</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>Feature</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tes_id</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>tss</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>tes</th>\n",
       "      <th>Frame</th>\n",
       "      <th>Start</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>ic_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ic</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>221624</th>\n",
       "      <td>ENSG00000000419[1,1,1]</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>ENSG00000000419_1</td>\n",
       "      <td>ENSG00000000419_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>50958580</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000000419.14[1,1,1]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>.</td>\n",
       "      <td>50934856</td>\n",
       "      <td>chr20</td>\n",
       "      <td>ENSG00000000419_1</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221623</th>\n",
       "      <td>ENSG00000000419[1,1,1]</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>ENSG00000000419_1</td>\n",
       "      <td>ENSG00000000419_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>50958580</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000000419.14[1,1,1]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>.</td>\n",
       "      <td>50934856</td>\n",
       "      <td>chr20</td>\n",
       "      <td>ENSG00000000419_1</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221644</th>\n",
       "      <td>ENSG00000000419[1,14,1]</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>ENSG00000000419_1</td>\n",
       "      <td>ENSG00000000419_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>50958580</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000000419.14[1,14,1]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>.</td>\n",
       "      <td>50934856</td>\n",
       "      <td>chr20</td>\n",
       "      <td>ENSG00000000419_14</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221643</th>\n",
       "      <td>ENSG00000000419[1,14,1]</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>ENSG00000000419_1</td>\n",
       "      <td>ENSG00000000419_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>50958580</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000000419.14[1,14,1]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>.</td>\n",
       "      <td>50934856</td>\n",
       "      <td>chr20</td>\n",
       "      <td>ENSG00000000419_14</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221660</th>\n",
       "      <td>ENSG00000000419[1,16,1]</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>ENSG00000000419_1</td>\n",
       "      <td>ENSG00000000419_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>50958580</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000000419.14[1,16,1]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>.</td>\n",
       "      <td>50934856</td>\n",
       "      <td>chr20</td>\n",
       "      <td>ENSG00000000419_16</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  transcript_id Score       gene_type     Feature  \\\n",
       "221624   ENSG00000000419[1,1,1]     .  protein_coding  transcript   \n",
       "221623   ENSG00000000419[1,1,1]     .  protein_coding  transcript   \n",
       "221644  ENSG00000000419[1,14,1]     .  protein_coding  transcript   \n",
       "221643  ENSG00000000419[1,14,1]     .  protein_coding  transcript   \n",
       "221660  ENSG00000000419[1,16,1]     .  protein_coding  transcript   \n",
       "\n",
       "                   gene_id             tes_id             tss_id  tss  \\\n",
       "221624  ENSG00000000419.14  ENSG00000000419_1  ENSG00000000419_1  1.0   \n",
       "221623  ENSG00000000419.14  ENSG00000000419_1  ENSG00000000419_1  1.0   \n",
       "221644  ENSG00000000419.14  ENSG00000000419_1  ENSG00000000419_1  1.0   \n",
       "221643  ENSG00000000419.14  ENSG00000000419_1  ENSG00000000419_1  1.0   \n",
       "221660  ENSG00000000419.14  ENSG00000000419_1  ENSG00000000419_1  1.0   \n",
       "\n",
       "             End Strand             transcript_name  tes Frame     Start  \\\n",
       "221624  50958580      -   ENSG00000000419.14[1,1,1]  1.0     .  50934856   \n",
       "221623  50958580      -   ENSG00000000419.14[1,1,1]  1.0     .  50934856   \n",
       "221644  50958580      -  ENSG00000000419.14[1,14,1]  1.0     .  50934856   \n",
       "221643  50958580      -  ENSG00000000419.14[1,14,1]  1.0     .  50934856   \n",
       "221660  50958580      -  ENSG00000000419.14[1,16,1]  1.0     .  50934856   \n",
       "\n",
       "       Chromosome               ic_id           gene_name    ic  \n",
       "221624      chr20   ENSG00000000419_1  ENSG00000000419.14   1.0  \n",
       "221623      chr20   ENSG00000000419_1  ENSG00000000419.14   1.0  \n",
       "221644      chr20  ENSG00000000419_14  ENSG00000000419.14  14.0  \n",
       "221643      chr20  ENSG00000000419_14  ENSG00000000419.14  14.0  \n",
       "221660      chr20  ENSG00000000419_16  ENSG00000000419.14  16.0  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# duplicated transcripts?\n",
    "gb_cols = ['Chromosome',\n",
    "             'Feature',\n",
    "             'Start', 'End',\n",
    "             'Score', 'Strand', 'Frame', 'gene_id', 'gene_name',\n",
    "             'gene_status', 'gene_type', 'talon_gene',\n",
    "             'ic', 'ic_id', 'tss_id', 'tss', 'tes_id', 'tes', 'transcript_id',\n",
    "             'transcript_name']\n",
    "gb_cols = list(set(df.columns)&set(gb_cols))\n",
    "temp = df.loc[df.Feature=='transcript']\n",
    "temp.loc[temp.transcript_id.duplicated(keep=False)].sort_values(by='transcript_id')[gb_cols].head()\n",
    "# assert len(temp.loc[temp.gene_id.duplicated()].index) == 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "926a3247-1017-435e-8af5-80b40f63e1e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42862\n",
      "57625\n",
      "42862\n"
     ]
    }
   ],
   "source": [
    "print(len(temp.transcript_id.unique()))\n",
    "print(len(temp.index))\n",
    "print(len(temp[gb_cols].drop_duplicates().index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "1f035d71-e95c-4e4e-bc8d-a845adf4ea21",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>ic_id</th>\n",
       "      <th>ic</th>\n",
       "      <th>Feature</th>\n",
       "      <th>tes_id</th>\n",
       "      <th>Start</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Score</th>\n",
       "      <th>tes</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>End</th>\n",
       "      <th>tss</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Frame</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>221624</th>\n",
       "      <td>chr20</td>\n",
       "      <td>ENSG00000000419.14[1,1,1]</td>\n",
       "      <td>ENSG00000000419_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENSG00000000419_1</td>\n",
       "      <td>50934856</td>\n",
       "      <td>ENSG00000000419[1,1,1]</td>\n",
       "      <td>.</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>50958580</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000000419_1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221623</th>\n",
       "      <td>chr20</td>\n",
       "      <td>ENSG00000000419.14[1,1,1]</td>\n",
       "      <td>ENSG00000000419_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENSG00000000419_1</td>\n",
       "      <td>50934856</td>\n",
       "      <td>ENSG00000000419[1,1,1]</td>\n",
       "      <td>.</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>50958580</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000000419_1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221644</th>\n",
       "      <td>chr20</td>\n",
       "      <td>ENSG00000000419.14[1,14,1]</td>\n",
       "      <td>ENSG00000000419_14</td>\n",
       "      <td>14.0</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENSG00000000419_1</td>\n",
       "      <td>50934856</td>\n",
       "      <td>ENSG00000000419[1,14,1]</td>\n",
       "      <td>.</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>50958580</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000000419_1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221643</th>\n",
       "      <td>chr20</td>\n",
       "      <td>ENSG00000000419.14[1,14,1]</td>\n",
       "      <td>ENSG00000000419_14</td>\n",
       "      <td>14.0</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENSG00000000419_1</td>\n",
       "      <td>50934856</td>\n",
       "      <td>ENSG00000000419[1,14,1]</td>\n",
       "      <td>.</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>50958580</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000000419_1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221660</th>\n",
       "      <td>chr20</td>\n",
       "      <td>ENSG00000000419.14[1,16,1]</td>\n",
       "      <td>ENSG00000000419_16</td>\n",
       "      <td>16.0</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENSG00000000419_1</td>\n",
       "      <td>50934856</td>\n",
       "      <td>ENSG00000000419[1,16,1]</td>\n",
       "      <td>.</td>\n",
       "      <td>1.0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>50958580</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000000419_1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome             transcript_name               ic_id    ic  \\\n",
       "221624      chr20   ENSG00000000419.14[1,1,1]   ENSG00000000419_1   1.0   \n",
       "221623      chr20   ENSG00000000419.14[1,1,1]   ENSG00000000419_1   1.0   \n",
       "221644      chr20  ENSG00000000419.14[1,14,1]  ENSG00000000419_14  14.0   \n",
       "221643      chr20  ENSG00000000419.14[1,14,1]  ENSG00000000419_14  14.0   \n",
       "221660      chr20  ENSG00000000419.14[1,16,1]  ENSG00000000419_16  16.0   \n",
       "\n",
       "           Feature             tes_id     Start            transcript_id  \\\n",
       "221624  transcript  ENSG00000000419_1  50934856   ENSG00000000419[1,1,1]   \n",
       "221623  transcript  ENSG00000000419_1  50934856   ENSG00000000419[1,1,1]   \n",
       "221644  transcript  ENSG00000000419_1  50934856  ENSG00000000419[1,14,1]   \n",
       "221643  transcript  ENSG00000000419_1  50934856  ENSG00000000419[1,14,1]   \n",
       "221660  transcript  ENSG00000000419_1  50934856  ENSG00000000419[1,16,1]   \n",
       "\n",
       "       Score  tes       gene_type       End  tss             tss_id Strand  \\\n",
       "221624     .  1.0  protein_coding  50958580  1.0  ENSG00000000419_1      -   \n",
       "221623     .  1.0  protein_coding  50958580  1.0  ENSG00000000419_1      -   \n",
       "221644     .  1.0  protein_coding  50958580  1.0  ENSG00000000419_1      -   \n",
       "221643     .  1.0  protein_coding  50958580  1.0  ENSG00000000419_1      -   \n",
       "221660     .  1.0  protein_coding  50958580  1.0  ENSG00000000419_1      -   \n",
       "\n",
       "                 gene_name             gene_id Frame  \n",
       "221624  ENSG00000000419.14  ENSG00000000419.14     .  \n",
       "221623  ENSG00000000419.14  ENSG00000000419.14     .  \n",
       "221644  ENSG00000000419.14  ENSG00000000419.14     .  \n",
       "221643  ENSG00000000419.14  ENSG00000000419.14     .  \n",
       "221660  ENSG00000000419.14  ENSG00000000419.14     .  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# duplicated transcripts?\n",
    "temp = df.loc[df.Feature=='transcript']\n",
    "temp.loc[temp.transcript_id.duplicated(keep=False)].sort_values(by='transcript_id')[gb_cols].head()\n",
    "# assert len(temp.loc[temp.transcript_id.duplicated()].index) == 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "10fd6192-1c84-412c-a1f4-f6e590f9f846",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(df.loc[df.duplicated(subset=gb_cols)])\n",
    "df = df.drop_duplicates(subset=gb_cols, keep='first')\n",
    "df = cerberus.sort_gtf(df)\n",
    "df = pr.PyRanges(df)\n",
    "df.to_gtf('test_merged.gtf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "26e49928-6768-4f94-84c3-50ea04caa200",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = cerberus.sort_gtf(df)\n",
    "df = pr.PyRanges(df)\n",
    "df.to_gtf('test_merged.gtf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "af91468c-a73b-496c-9d24-d56628cca095",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.df\n",
    "df = df.loc[df.Feature.isin(['transcript', 'exon'])]\n",
    "df = pr.PyRanges(df)\n",
    "df.to_gtf('test_merged_no_genes.gtf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "c39b4137-4e9b-4710-9e7a-ddff2cad1ce1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "42862"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.df\n",
    "len(df.loc[df.Feature=='transcript'].index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2b450da0-05f9-47f3-b3e6-c325bef73034",
   "metadata": {},
   "outputs": [],
   "source": [
    "gb_cols = ['Chromosome',\n",
    "                 'Feature',\n",
    "                 'Start', 'End',\n",
    "                 'Score', 'Strand', 'Frame', 'gene_id', 'gene_name',\n",
    "                 'gene_status', 'gene_type', 'talon_gene',\n",
    "                 'ic', 'ic_id', 'tss_id', 'tss', 'tes_id', 'tes', 'transcript_id',\n",
    "                 'transcript_name']\n",
    "gb_cols = list(set(temp.columns)&set(gb_cols))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "65fcc804-9217-4a25-9af7-58bb48ec1b7f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "67390\n",
      "44489\n"
     ]
    }
   ],
   "source": [
    "temp = temp[gb_cols]\n",
    "print(len(temp.index))\n",
    "temp.drop_duplicates(inplace=True)\n",
    "print(len(temp.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "9e89f488-773b-4645-8dd8-599a60923744",
   "metadata": {},
   "outputs": [],
   "source": [
    "def collapse_non_gb_col(x):\n",
    "    x = x.fillna('')\n",
    "    x = x.astype(str)\n",
    "    x = x.unique().tolist()\n",
    "    x = ','.join(x)\n",
    "    if x == '':\n",
    "        x = np.nan\n",
    "    return(x)\n",
    "\n",
    "gb_cols = ['Chromosome',\n",
    "             'Feature',\n",
    "             'Start', 'End',\n",
    "             'Score', 'Strand', 'Frame', 'gene_id', 'gene_name',\n",
    "             'gene_status', 'gene_type', 'talon_gene',\n",
    "             'ic', 'ic_id', 'tss_id', 'tss', 'tes_id', 'tes', 'transcript_id',\n",
    "             'transcript_name']\n",
    "gb_cols = list(set(df.columns)&set(gb_cols))\n",
    "\n",
    "# get dictionary:function mapping for aggregation\n",
    "agg_cols = list(set(df.columns)-set(gb_cols))\n",
    "agg_dict = dict()\n",
    "for c in agg_cols:\n",
    "    agg_dict[c] = collapse_non_gb_col\n",
    "\n",
    "# get collapsed features to add to deduplicated df\n",
    "t_df = df.loc[df.Feature == 'transcript'].copy(deep=True)\n",
    "\n",
    "non_gb_cols = list(set(t_df.columns.tolist())-set(gb_cols))\n",
    "t_df[non_gb_cols].fillna('', inplace=True)\n",
    "t_df = t_df.groupby(gb_cols, observed=True).agg(agg_dict).reset_index()\n",
    "t_df = t_df[['transcript_id']+non_gb_cols]\n",
    "collapsed_feats = t_df.copy(deep=True)\n",
    "\n",
    "# # replace the non gb columns with the ones that we already grouped\n",
    "# df.drop(non_gb_cols, axis=1, inplace=True)\n",
    "# df = df.merge(collapsed_feats, how='left', on='transcript_id'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "659dbe1e-da0b-40d6-b946-3797d5ea42de",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42103\n",
      "42103\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Source</th>\n",
       "      <th>tes_last_sa_issue</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>tss_first_sd_issue</th>\n",
       "      <th>original_transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003[1,1,1]</td>\n",
       "      <td>annotated_isoform</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000373020.9</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000373020.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000419[1,1,1]</td>\n",
       "      <td>annotated_isoform</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000371588.10</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000371588.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419[1,10,10]</td>\n",
       "      <td>annotated_isoform</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000684628.1</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000684628.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457[1,1,1]</td>\n",
       "      <td>annotated_isoform</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000367771.11</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000367771.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000457[3,3,2]</td>\n",
       "      <td>annotated_isoform</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000367772.8</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000367772.8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              transcript_id             Source tes_last_sa_issue  \\\n",
       "0    ENSG00000000003[1,1,1]  annotated_isoform             False   \n",
       "1    ENSG00000000419[1,1,1]  annotated_isoform             False   \n",
       "2  ENSG00000000419[1,10,10]  annotated_isoform             False   \n",
       "3    ENSG00000000457[1,1,1]  annotated_isoform             False   \n",
       "4    ENSG00000000457[3,3,2]  annotated_isoform             False   \n",
       "\n",
       "  original_transcript_name tss_first_sd_issue original_transcript_id  \n",
       "0        ENST00000373020.9              False      ENST00000373020.9  \n",
       "1       ENST00000371588.10              False     ENST00000371588.10  \n",
       "2        ENST00000684628.1              False      ENST00000684628.1  \n",
       "3       ENST00000367771.11              False     ENST00000367771.11  \n",
       "4        ENST00000367772.8              False      ENST00000367772.8  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(len(t_df.index))\n",
    "print(len(t_df.transcript_id.unique()))\n",
    "t_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "91a4496f-948f-491b-87a9-23b175306a6c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42103\n",
      "42103\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Source</th>\n",
       "      <th>tes_last_sa_issue</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>tss_first_sd_issue</th>\n",
       "      <th>original_transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003[1,1,1]</td>\n",
       "      <td>annotated_isoform</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000373020.9</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000373020.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000419[1,1,1]</td>\n",
       "      <td>annotated_isoform</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000371588.10</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000371588.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419[1,10,10]</td>\n",
       "      <td>annotated_isoform</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000684628.1</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000684628.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457[1,1,1]</td>\n",
       "      <td>annotated_isoform</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000367771.11</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000367771.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000457[3,3,2]</td>\n",
       "      <td>annotated_isoform</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000367772.8</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000367772.8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              transcript_id             Source tes_last_sa_issue  \\\n",
       "0    ENSG00000000003[1,1,1]  annotated_isoform             False   \n",
       "1    ENSG00000000419[1,1,1]  annotated_isoform             False   \n",
       "2  ENSG00000000419[1,10,10]  annotated_isoform             False   \n",
       "3    ENSG00000000457[1,1,1]  annotated_isoform             False   \n",
       "4    ENSG00000000457[3,3,2]  annotated_isoform             False   \n",
       "\n",
       "  original_transcript_name tss_first_sd_issue original_transcript_id  \n",
       "0        ENST00000373020.9              False      ENST00000373020.9  \n",
       "1       ENST00000371588.10              False     ENST00000371588.10  \n",
       "2        ENST00000684628.1              False      ENST00000684628.1  \n",
       "3       ENST00000367771.11              False     ENST00000367771.11  \n",
       "4        ENST00000367772.8              False      ENST00000367772.8  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(len(collapsed_feats.index))\n",
    "print(len(collapsed_feats.transcript_id.unique()))\n",
    "collapsed_feats.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "3b7d212f-c5ae-4087-b9fc-c989cc8decf2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42104\n"
     ]
    }
   ],
   "source": [
    "# deduplicate df based only on transcript id\n",
    "temp = df[['transcript_id', 'original_transcript_id']].drop_duplicates()\n",
    "print(len(temp.index))\n",
    "dupe_old_tids = temp.loc[temp.transcript_id.duplicated(keep='first'), 'original_transcript_id']\n",
    "df = df.loc[~df.original_transcript_id.isin(dupe_old_tids)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "da318c68-91c8-4d1e-8b61-e2e159dc576e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>original_transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [transcript_id, original_transcript_id]\n",
       "Index: []"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.loc[temp.transcript_id.duplicated()].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "3a072cb5-d42c-4c32-8232-cec0e106b203",
   "metadata": {},
   "outputs": [],
   "source": [
    "# temp2 = temp[['transcript_id', 'original_transcript_id']].merge(t_df[['transcript_id', 'original_transcript_id', 'Source']],\n",
    "#                                                                 how='outer', on=['transcript_id', 'original_transcript_id'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "05be31f6-f0ac-4e4b-be1c-d660f0dd6add",
   "metadata": {},
   "outputs": [],
   "source": [
    "# temp2.loc[temp2.Source.isnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eeb8c692-a152-4bf1-bec1-e55d22eb08a1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
