{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f9499390-7181-46d6-944b-1785b4ca025d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import  cerberus\n",
    "import glob\n",
    "import pyranges as pr\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "dadd8446-4d42-45c7-b390-7c2f17485859",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame()\n",
    "for f in glob.glob('*gtf'):\n",
    "    df = pd.concat([pr.read_gtf(f, as_df=True), df], axis=0)\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "2ae54b0c-d2db-4aef-9094-6d25f56804a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>ic</th>\n",
       "      <th>...</th>\n",
       "      <th>tss</th>\n",
       "      <th>tes_id</th>\n",
       "      <th>tes</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>tes_last_sa_issue</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>tss_first_sd_issue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>GL000195.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript</td>\n",
       "      <td>48901</td>\n",
       "      <td>86687</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227</th>\n",
       "      <td>GL000205.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript</td>\n",
       "      <td>137791</td>\n",
       "      <td>145075</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>GL000219.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript</td>\n",
       "      <td>98559</td>\n",
       "      <td>99612</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11357</th>\n",
       "      <td>chr1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript</td>\n",
       "      <td>70205727</td>\n",
       "      <td>70208101</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11360</th>\n",
       "      <td>chr1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript</td>\n",
       "      <td>70205732</td>\n",
       "      <td>70215360</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome Source     Feature     Start       End Score Strand Frame  \\\n",
       "114    GL000195.1    NaN  transcript     48901     86687     .      -     .   \n",
       "227    GL000205.2    NaN  transcript    137791    145075     .      -     .   \n",
       "282    GL000219.1    NaN  transcript     98559     99612     .      +     .   \n",
       "11357        chr1    NaN  transcript  70205727  70208101     .      +     .   \n",
       "11360        chr1    NaN  transcript  70205732  70215360     .      +     .   \n",
       "\n",
       "      gene_id   ic  ...  tss tes_id  tes transcript_id transcript_name  \\\n",
       "114        NA  NaN  ...  NaN    NaN  NaN           NaN             NaN   \n",
       "227        NA  NaN  ...  NaN    NaN  NaN           NaN             NaN   \n",
       "282        NA  NaN  ...  NaN    NaN  NaN           NaN             NaN   \n",
       "11357      NA  NaN  ...  NaN    NaN  NaN           NaN             NaN   \n",
       "11360      NA  NaN  ...  NaN    NaN  NaN           NaN             NaN   \n",
       "\n",
       "      gene_name tes_last_sa_issue original_transcript_id  \\\n",
       "114         NaN               NaN                    NaN   \n",
       "227         NaN               NaN                    NaN   \n",
       "282         NaN               NaN                    NaN   \n",
       "11357       NaN               NaN                    NaN   \n",
       "11360       NaN               NaN                    NaN   \n",
       "\n",
       "      original_transcript_name tss_first_sd_issue  \n",
       "114                        NaN                NaN  \n",
       "227                        NaN                NaN  \n",
       "282                        NaN                NaN  \n",
       "11357                      NaN                NaN  \n",
       "11360                      NaN                NaN  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = df.loc[df.Feature=='transcript']\n",
    "temp.loc[temp.transcript_id.duplicated(keep=False)].sort_values(by='transcript_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "dfdf012e-cae7-4acd-8976-d3fe59090cd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = cerberus.sort_gtf(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7e9a1c68-7488-43ad-8208-4f0422455c65",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "401447\n",
      "399412\n"
     ]
    }
   ],
   "source": [
    "print(len(df.index))\n",
    "df = cerberus.agg_gtf(df)\n",
    "print(len(df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "5f805394-10fc-400b-86ee-fa5c0fec4840",
   "metadata": {},
   "outputs": [],
   "source": [
    "# duplicated genes?\n",
    "temp = df.loc[df.Feature=='gene']\n",
    "temp.loc[temp.gene_id.duplicated(keep=False)].sort_values(by='gene_id').head()\n",
    "assert len(temp.loc[temp.gene_id.duplicated()].index) == 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "1f035d71-e95c-4e4e-bc8d-a845adf4ea21",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>ic</th>\n",
       "      <th>ic_id</th>\n",
       "      <th>...</th>\n",
       "      <th>tes_id</th>\n",
       "      <th>tes</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>tes_last_sa_issue</th>\n",
       "      <th>Source</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>tss_first_sd_issue</th>\n",
       "      <th>original_transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>199754</th>\n",
       "      <td>chrX</td>\n",
       "      <td>transcript</td>\n",
       "      <td>100627097</td>\n",
       "      <td>100636831</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000003.16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000000003_1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000000003_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000000003[1,1,1]</td>\n",
       "      <td>ENSG00000000003[1,1,1]</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>False</td>\n",
       "      <td>annotated_isoform</td>\n",
       "      <td>ENST00000373020.9</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000373020.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199753</th>\n",
       "      <td>chrX</td>\n",
       "      <td>transcript</td>\n",
       "      <td>100627097</td>\n",
       "      <td>100636831</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000003.16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000000003_1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000000003_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000000003[1,1,1]</td>\n",
       "      <td>ENSG00000000003[1,1,1]</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>False</td>\n",
       "      <td>annotated_isoform</td>\n",
       "      <td>ENST00000373020.9</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000373020.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>169795014</td>\n",
       "      <td>169804357</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>4.0</td>\n",
       "      <td>ENSG00000000460_4</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000000460_2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000000460[1,4,2]</td>\n",
       "      <td>ENSG00000000460[1,4,2]</td>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>False</td>\n",
       "      <td>annotated_isoform</td>\n",
       "      <td>ENST00000496973.5</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000496973.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>169795014</td>\n",
       "      <td>169804357</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>4.0</td>\n",
       "      <td>ENSG00000000460_4</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000000460_2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000000460[1,4,2]</td>\n",
       "      <td>ENSG00000000460[1,4,2]</td>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>False</td>\n",
       "      <td>annotated_isoform</td>\n",
       "      <td>ENST00000496973.5</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000496973.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199828</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>27612281</td>\n",
       "      <td>27626265</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000000938.13</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSG00000000938_3</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000000938_2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000000938[3,3,2]</td>\n",
       "      <td>ENSG00000000938[3,3,2]</td>\n",
       "      <td>ENSG00000000938</td>\n",
       "      <td>False</td>\n",
       "      <td>annotated_isoform</td>\n",
       "      <td>ENST00000399173.5</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000399173.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395602</th>\n",
       "      <td>chr9</td>\n",
       "      <td>transcript</td>\n",
       "      <td>24329</td>\n",
       "      <td>25007</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395603</th>\n",
       "      <td>chr9</td>\n",
       "      <td>transcript</td>\n",
       "      <td>24329</td>\n",
       "      <td>25007</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395604</th>\n",
       "      <td>chr9</td>\n",
       "      <td>transcript</td>\n",
       "      <td>24328</td>\n",
       "      <td>25007</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395605</th>\n",
       "      <td>chr9</td>\n",
       "      <td>transcript</td>\n",
       "      <td>23296</td>\n",
       "      <td>25007</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395606</th>\n",
       "      <td>chr9</td>\n",
       "      <td>transcript</td>\n",
       "      <td>20189</td>\n",
       "      <td>25007</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>48156 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome     Feature      Start        End Score Strand Frame  \\\n",
       "199754       chrX  transcript  100627097  100636831     .      -     .   \n",
       "199753       chrX  transcript  100627097  100636831     .      -     .   \n",
       "0            chr1  transcript  169795014  169804357     .      +     .   \n",
       "1            chr1  transcript  169795014  169804357     .      +     .   \n",
       "199828       chr1  transcript   27612281   27626265     .      -     .   \n",
       "...           ...         ...        ...        ...   ...    ...   ...   \n",
       "395602       chr9  transcript      24329      25007     .      -     .   \n",
       "395603       chr9  transcript      24329      25007     .      -     .   \n",
       "395604       chr9  transcript      24328      25007     .      -     .   \n",
       "395605       chr9  transcript      23296      25007     .      -     .   \n",
       "395606       chr9  transcript      20189      25007     .      -     .   \n",
       "\n",
       "                   gene_id   ic              ic_id  ...             tes_id  \\\n",
       "199754  ENSG00000000003.16  1.0  ENSG00000000003_1  ...  ENSG00000000003_1   \n",
       "199753  ENSG00000000003.16  1.0  ENSG00000000003_1  ...  ENSG00000000003_1   \n",
       "0       ENSG00000000460.17  4.0  ENSG00000000460_4  ...  ENSG00000000460_2   \n",
       "1       ENSG00000000460.17  4.0  ENSG00000000460_4  ...  ENSG00000000460_2   \n",
       "199828  ENSG00000000938.13  3.0  ENSG00000000938_3  ...  ENSG00000000938_2   \n",
       "...                    ...  ...                ...  ...                ...   \n",
       "395602                  NA  NaN                NaN  ...                NaN   \n",
       "395603                  NA  NaN                NaN  ...                NaN   \n",
       "395604                  NA  NaN                NaN  ...                NaN   \n",
       "395605                  NA  NaN                NaN  ...                NaN   \n",
       "395606                  NA  NaN                NaN  ...                NaN   \n",
       "\n",
       "        tes           transcript_id         transcript_name        gene_name  \\\n",
       "199754  1.0  ENSG00000000003[1,1,1]  ENSG00000000003[1,1,1]  ENSG00000000003   \n",
       "199753  1.0  ENSG00000000003[1,1,1]  ENSG00000000003[1,1,1]  ENSG00000000003   \n",
       "0       2.0  ENSG00000000460[1,4,2]  ENSG00000000460[1,4,2]  ENSG00000000460   \n",
       "1       2.0  ENSG00000000460[1,4,2]  ENSG00000000460[1,4,2]  ENSG00000000460   \n",
       "199828  2.0  ENSG00000000938[3,3,2]  ENSG00000000938[3,3,2]  ENSG00000000938   \n",
       "...     ...                     ...                     ...              ...   \n",
       "395602  NaN                     NaN                     NaN              NaN   \n",
       "395603  NaN                     NaN                     NaN              NaN   \n",
       "395604  NaN                     NaN                     NaN              NaN   \n",
       "395605  NaN                     NaN                     NaN              NaN   \n",
       "395606  NaN                     NaN                     NaN              NaN   \n",
       "\n",
       "       tes_last_sa_issue             Source original_transcript_name  \\\n",
       "199754             False  annotated_isoform        ENST00000373020.9   \n",
       "199753             False  annotated_isoform        ENST00000373020.9   \n",
       "0                  False  annotated_isoform        ENST00000496973.5   \n",
       "1                  False  annotated_isoform        ENST00000496973.5   \n",
       "199828             False  annotated_isoform        ENST00000399173.5   \n",
       "...                  ...                ...                      ...   \n",
       "395602               NaN                NaN                      NaN   \n",
       "395603               NaN                NaN                      NaN   \n",
       "395604               NaN                NaN                      NaN   \n",
       "395605               NaN                NaN                      NaN   \n",
       "395606               NaN                NaN                      NaN   \n",
       "\n",
       "       tss_first_sd_issue original_transcript_id  \n",
       "199754              False      ENST00000373020.9  \n",
       "199753              False      ENST00000373020.9  \n",
       "0                   False      ENST00000496973.5  \n",
       "1                   False      ENST00000496973.5  \n",
       "199828              False      ENST00000399173.5  \n",
       "...                   ...                    ...  \n",
       "395602                NaN                    NaN  \n",
       "395603                NaN                    NaN  \n",
       "395604                NaN                    NaN  \n",
       "395605                NaN                    NaN  \n",
       "395606                NaN                    NaN  \n",
       "\n",
       "[48156 rows x 22 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# duplicated transcripts?\n",
    "temp = df.loc[df.Feature=='transcript']\n",
    "temp.loc[temp.transcript_id.duplicated(keep=False)].sort_values(by='transcript_id')\n",
    "# assert len(temp.loc[temp.transcript_id.duplicated()].index) == 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2b450da0-05f9-47f3-b3e6-c325bef73034",
   "metadata": {},
   "outputs": [],
   "source": [
    "gb_cols = ['Chromosome',\n",
    "                 'Feature',\n",
    "                 'Start', 'End',\n",
    "                 'Score', 'Strand', 'Frame', 'gene_id', 'gene_name',\n",
    "                 'gene_status', 'gene_type', 'talon_gene',\n",
    "                 'ic', 'ic_id', 'tss_id', 'tss', 'tes_id', 'tes', 'transcript_id',\n",
    "                 'transcript_name']\n",
    "gb_cols = list(set(temp.columns)&set(gb_cols))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "65fcc804-9217-4a25-9af7-58bb48ec1b7f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "67390\n",
      "44489\n"
     ]
    }
   ],
   "source": [
    "temp = temp[gb_cols]\n",
    "print(len(temp.index))\n",
    "temp.drop_duplicates(inplace=True)\n",
    "print(len(temp.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "9e89f488-773b-4645-8dd8-599a60923744",
   "metadata": {},
   "outputs": [],
   "source": [
    "def collapse_non_gb_col(x):\n",
    "    x = x.fillna('')\n",
    "    x = x.astype(str)\n",
    "    x = x.unique().tolist()\n",
    "    x = ','.join(x)\n",
    "    if x == '':\n",
    "        x = np.nan\n",
    "    return(x)\n",
    "\n",
    "gb_cols = ['Chromosome',\n",
    "             'Feature',\n",
    "             'Start', 'End',\n",
    "             'Score', 'Strand', 'Frame', 'gene_id', 'gene_name',\n",
    "             'gene_status', 'gene_type', 'talon_gene',\n",
    "             'ic', 'ic_id', 'tss_id', 'tss', 'tes_id', 'tes', 'transcript_id',\n",
    "             'transcript_name']\n",
    "gb_cols = list(set(df.columns)&set(gb_cols))\n",
    "\n",
    "# get dictionary:function mapping for aggregation\n",
    "agg_cols = list(set(df.columns)-set(gb_cols))\n",
    "agg_dict = dict()\n",
    "for c in agg_cols:\n",
    "    agg_dict[c] = collapse_non_gb_col\n",
    "\n",
    "# get collapsed features to add to deduplicated df\n",
    "t_df = df.loc[df.Feature == 'transcript'].copy(deep=True)\n",
    "\n",
    "non_gb_cols = list(set(t_df.columns.tolist())-set(gb_cols))\n",
    "t_df[non_gb_cols].fillna('', inplace=True)\n",
    "t_df = t_df.groupby(gb_cols, observed=True).agg(agg_dict).reset_index()\n",
    "t_df = t_df[['transcript_id']+non_gb_cols]\n",
    "collapsed_feats = t_df.copy(deep=True)\n",
    "\n",
    "# # replace the non gb columns with the ones that we already grouped\n",
    "# df.drop(non_gb_cols, axis=1, inplace=True)\n",
    "# df = df.merge(collapsed_feats, how='left', on='transcript_id'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "659dbe1e-da0b-40d6-b946-3797d5ea42de",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42103\n",
      "42103\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Source</th>\n",
       "      <th>tes_last_sa_issue</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>tss_first_sd_issue</th>\n",
       "      <th>original_transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003[1,1,1]</td>\n",
       "      <td>annotated_isoform</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000373020.9</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000373020.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000419[1,1,1]</td>\n",
       "      <td>annotated_isoform</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000371588.10</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000371588.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419[1,10,10]</td>\n",
       "      <td>annotated_isoform</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000684628.1</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000684628.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457[1,1,1]</td>\n",
       "      <td>annotated_isoform</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000367771.11</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000367771.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000457[3,3,2]</td>\n",
       "      <td>annotated_isoform</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000367772.8</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000367772.8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              transcript_id             Source tes_last_sa_issue  \\\n",
       "0    ENSG00000000003[1,1,1]  annotated_isoform             False   \n",
       "1    ENSG00000000419[1,1,1]  annotated_isoform             False   \n",
       "2  ENSG00000000419[1,10,10]  annotated_isoform             False   \n",
       "3    ENSG00000000457[1,1,1]  annotated_isoform             False   \n",
       "4    ENSG00000000457[3,3,2]  annotated_isoform             False   \n",
       "\n",
       "  original_transcript_name tss_first_sd_issue original_transcript_id  \n",
       "0        ENST00000373020.9              False      ENST00000373020.9  \n",
       "1       ENST00000371588.10              False     ENST00000371588.10  \n",
       "2        ENST00000684628.1              False      ENST00000684628.1  \n",
       "3       ENST00000367771.11              False     ENST00000367771.11  \n",
       "4        ENST00000367772.8              False      ENST00000367772.8  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(len(t_df.index))\n",
    "print(len(t_df.transcript_id.unique()))\n",
    "t_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "91a4496f-948f-491b-87a9-23b175306a6c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42103\n",
      "42103\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Source</th>\n",
       "      <th>tes_last_sa_issue</th>\n",
       "      <th>original_transcript_name</th>\n",
       "      <th>tss_first_sd_issue</th>\n",
       "      <th>original_transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003[1,1,1]</td>\n",
       "      <td>annotated_isoform</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000373020.9</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000373020.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000419[1,1,1]</td>\n",
       "      <td>annotated_isoform</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000371588.10</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000371588.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419[1,10,10]</td>\n",
       "      <td>annotated_isoform</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000684628.1</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000684628.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457[1,1,1]</td>\n",
       "      <td>annotated_isoform</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000367771.11</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000367771.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000457[3,3,2]</td>\n",
       "      <td>annotated_isoform</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000367772.8</td>\n",
       "      <td>False</td>\n",
       "      <td>ENST00000367772.8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              transcript_id             Source tes_last_sa_issue  \\\n",
       "0    ENSG00000000003[1,1,1]  annotated_isoform             False   \n",
       "1    ENSG00000000419[1,1,1]  annotated_isoform             False   \n",
       "2  ENSG00000000419[1,10,10]  annotated_isoform             False   \n",
       "3    ENSG00000000457[1,1,1]  annotated_isoform             False   \n",
       "4    ENSG00000000457[3,3,2]  annotated_isoform             False   \n",
       "\n",
       "  original_transcript_name tss_first_sd_issue original_transcript_id  \n",
       "0        ENST00000373020.9              False      ENST00000373020.9  \n",
       "1       ENST00000371588.10              False     ENST00000371588.10  \n",
       "2        ENST00000684628.1              False      ENST00000684628.1  \n",
       "3       ENST00000367771.11              False     ENST00000367771.11  \n",
       "4        ENST00000367772.8              False      ENST00000367772.8  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(len(collapsed_feats.index))\n",
    "print(len(collapsed_feats.transcript_id.unique()))\n",
    "collapsed_feats.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "3b7d212f-c5ae-4087-b9fc-c989cc8decf2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42104\n"
     ]
    }
   ],
   "source": [
    "# deduplicate df based only on transcript id\n",
    "temp = df[['transcript_id', 'original_transcript_id']].drop_duplicates()\n",
    "print(len(temp.index))\n",
    "dupe_old_tids = temp.loc[temp.transcript_id.duplicated(keep='first'), 'original_transcript_id']\n",
    "df = df.loc[~df.original_transcript_id.isin(dupe_old_tids)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "da318c68-91c8-4d1e-8b61-e2e159dc576e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>original_transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [transcript_id, original_transcript_id]\n",
       "Index: []"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.loc[temp.transcript_id.duplicated()].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "3a072cb5-d42c-4c32-8232-cec0e106b203",
   "metadata": {},
   "outputs": [],
   "source": [
    "# temp2 = temp[['transcript_id', 'original_transcript_id']].merge(t_df[['transcript_id', 'original_transcript_id', 'Source']],\n",
    "#                                                                 how='outer', on=['transcript_id', 'original_transcript_id'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "05be31f6-f0ac-4e4b-be1c-d660f0dd6add",
   "metadata": {},
   "outputs": [],
   "source": [
    "# temp2.loc[temp2.Source.isnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eeb8c692-a152-4bf1-bec1-e55d22eb08a1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
